---
title: "Simulations in population genetics"
subtitle: "[PopGen summer course 2023](http://www.popgen.dk/popgen23/)"
author:
  - "Martin Petr"
  - "[CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)"
date: "August 2023"
date-format: "MMMM YYYY"
format:
  revealjs:
    echo: true
    code-line-numbers: false
    fig-align: center
    slide-number: true
  # html:
  #   echo: true
  #   code-line-numbers: false
  #   fig-align: center
---

# 

::: columns
::: {.column width="70%"}
> Many problems in population genetics cannot be solved by a mathematician, no matter how gifted. \[It\] is already clear that computer methods are very powerful. This is good. It \[...\] **permits people with limited mathematical knowledge to work on important problems** \[...\].
:::

::: {.column width="30%"}
![](images/crow.jpeg)

[James F. Crow](https://en.wikipedia.org/wiki/James_F._Crow) -- [interview](http://www.gnxp.com/blog/2006/06/10-questions-for-jim-crow.php)
:::
:::

# Why use simulations?

1.  Making sense of estimated statistics
2.  Fitting model parameters (i.e. [ABC](https://en.wikipedia.org/wiki/Approximate_Bayesian_computation))
3.  Ground truth for method development

## Making sense of estimated statistics

<center>![](images/fstats_sims.png)</center>

::: aside
Image from [Peter (2016)](https://academic.oup.com/genetics/article/202/4/1485/5930214)
:::

## Fitting model parameters (i.e. [ABC](https://en.wikipedia.org/wiki/Approximate_Bayesian_computation))

<center>![](images/abc_scheme.png){width="50%"}</center>

::: aside
Image from [Wikipedia on ABC](https://en.wikipedia.org/wiki/Approximate_Bayesian_computation)
:::

## Ground truth for method development

<center>![](images/mcmc.png)</center>

::: aside
Image from [Schiffels and Durbin (2014)](https://www.nature.com/articles/ng.3015)
:::

# What does it mean to simulate a genome?

<br> How would you design an algorithm for a popgen simulation?

</h2>

<br>What minimum components are needed?

# If we want to simulate population genetics

<br>

. . .

We need *populations*.

. . .

We need *genetics*.

# A chromosome is...

<br>

. . .

...a linear sequence of nucleotides

-   a list of characters (A/G/C/T nucleotides)
-   a list of 0 or 1 values (ancestral/derived allele),

. . .

which mutates at a given *mutation rate*,<br>

and recombines at a certain *recombination rate*.

# A population is...

. . . 

<br>

... a collection of *individuals* at a given point in time,<br><br>
each carrying *chromosomes* inherited from its parents.

# 

<h1>Homebrewed<br>single-locus<br>simulation in R</h1>

<br>

<h3>"What is the expected trajectory of a (neutral) allele under the influence of genetic drift?"</h3>

## Complete simulation R script

```{r}
#| results: hide
#| code-line-numbers: "|1-3|5-6|8-9|10-11|13-14|17"
p_0 <- 0.5  # initial allele frequency
N   <- 500  # number of chromosomes in a population
T   <- 500  # number of generations to simulate

# a vector for storing frequencies over time
p_trajectory <- p_0 

# in each generation:
for (gen_i in 2:T) {
  # get frequency in the previous generation
  p_prev <- p_trajectory[gen_i - 1] 

  # calculate frequency in this generation
  p_trajectory[gen_i] <- rbinom(1, N, p_prev) / N 
}

p_trajectory
```

## $N$ = 500, $p_0 = 0.5$

```{r}
#| code-fold: true
plot(p_trajectory, type = "l", ylim = c(0, 1),
     xlab = "generations", ylab = "allele frequency")
abline(h = p_0, lty = 2, col = "red")
```

## Let's make it a function

**Input:** $p_0$, $N$, and the number of generations

**Output:** allele frequency trajectory vector

```{r}
#| code-line-numbers: "|1"
simulate <- function(p_0, N, T) {
  p_trajectory <- p_0

  for (gen_i in 2:T) {
    p_prev <- p_trajectory[gen_i - 1]
    p <- rbinom(1, N, p_prev) / N
    p_trajectory[gen_i] <- p
  }

  p_trajectory
}
```

## $N$ = 500, $p_0 = 0.5$ (20 replicates)

```{r}
#| code-fold: true
#| fig-align: center
reps <- replicate(20, simulate(N = 500, p_0 = 0.5, T = 2000))

matplot(reps, ylim = c(0, 1), xlab = "generations", ylab = "allele frequency", type = "l", lty = 1)
```

## $N$ = 1000, $p_0 = 0.5$ (20 replicates)

```{r}
#| code-fold: true
#| fig-align: center
reps <- replicate(20, simulate(N = 1000, p_0 = 0.5, T = 2000))

matplot(reps, ylim = c(0, 1), xlab = "generations", ylab = "allele frequency", type = "l", lty = 1)
```

## $N$ = 5000, $p_0 = 0.5$ (20 replicates)

```{r}
#| code-fold: true
#| fig-align: center
reps <- replicate(20, simulate(N = 5000, p_0 = 0.5, T = 2000))

matplot(reps, ylim = c(0, 1), xlab = "generations", ylab = "allele frequency", type = "l", lty = 1)
```

## $N$ = 10000, $p_0 = 0.5$ (20 replicates)

```{r}
#| code-fold: true
#| fig-align: center
reps <- replicate(20, simulate(N = 10000, p_0 = 0.5, T = 30000))

matplot(reps, ylim = c(0, 1), xlab = "generations", ylab = "allele frequency", type = "l", lty = 1)
```

## $N$ = 10000, $p_0 = 0.5$ (20 replicates)

```{r}
#| code-fold: true
#| fig-align: center
factors <- MASS::fractions(c(3, 2, 1, 1/2, 1/5, 1/10))
matplot(reps, ylim = c(0, 1), xlab = "generations", ylab = "allele frequency", type = "l", lty = 1)
abline(v = 10000 * factors, lwd = 5)
```

## Expected allele frequency distribution

::: columns
::: {.column width="50%"}
```{r}
#| code-fold: true
#| fig-align: center
#| fig-height: 10
#| fig-width: 10
library(ggplot2)
library(dplyr)
library(parallel)

p_0 <- 0.5
N <- 10000
factors <- MASS::fractions(c(3, 2, 1, 1/2, 1/5, 1/10))

final_frequencies <- parallel::mclapply(
  seq_along(factors), function(i) {
    f <- factors[i]
    t <- as.integer(N * f)
    # get complete trajectories as a matrix (each column a single trajectory)
    reps <- replicate(1000, simulate(N = N, p_0 = p_0, T = t))
    # only keep the last slice of the matrix with the final frequencies
    data.frame(
      t = sprintf("t = %s * N", f),
      freq = reps[t, ]
    )
  },
  mc.cores = parallel::detectCores()
) %>% do.call(rbind, .)
final_frequencies$t <- forcats::fct_rev(forcats::fct_relevel(final_frequencies$t, sprintf("t = %s * N", factors)))

final_frequencies %>% .[.$freq > 0 & .$freq < 1, ] %>%
ggplot() +
  geom_histogram(aes(freq, y = after_stat(density), fill = t), position = "identity", bins = 100, alpha = 0.75) +
  labs(x = "allele frequency") +
  coord_cartesian(ylim = c(0, 3)) +
  facet_grid(t ~ .) +
  guides(fill = guide_legend(sprintf("time since\nthe start\n[assuming\nN = %s]", N))) +
  theme_minimal() +
  theme(strip.text.y = element_blank(),
        axis.title.x = element_text(size = 15),
        axis.title.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        axis.text.y = element_text(size = 15),
        legend.title = element_text(size = 15),
        legend.text = element_text(size = 15))
```
:::

::: {.column width="50%"}
::: fragment
![*"Diffusion Models in Population Genetics"*, [Kimura (1964)](https://www.jstor.org/stable/3211856#metadata_info_tab_contents)](images/kimura.jpeg)
:::
:::
:::

#  {background-image="images/montypython.jpeg"}

::: fragment

<h1><br><br><br></h1>

<h1 color="black" style="background-color: white">

But now for something completely different.

</h1>
:::

#  {background-image="images/montypython.jpeg"}

<h1><br><br><br></h1>

<h1 color="black" style="background-color: white">

Let's talk about<br>"real" simulations.

</h1>

## There are many simulation tools

<br>

The most famous and widely used are [SLiM](https://messerlab.org/slim/) and [*msprime*](https://tskit.dev/msprime/docs/stable/intro.html).

<br>

They are <u>very</u> powerful but both require:

::: incremental

- quite a bit of programming knowledge,
- a lot of code for non-trivial simulations (üêõü™≤üêú).

:::

::: fragment

<br>

<center><h3>**The exercises will focus on the [*slendr*](http://www.slendr.net)**<br>popgen simulation toolkit for R.</h3></center>
:::

## 

::: columns
::: {.column width="60%"}
<h2>What is SLiM?</h2>

::: fragment
-   **Forward-time simulator**

-   It's fully programmable!
:::

::: fragment
-   Massive library of functions for:
    -   Demographic events
    -   Various mating systems
    -   Selection, quantitative traits, ...
:::

::: fragment
-   \> 700 pages long [manual](https://github.com/MesserLab/SLiM/releases/download/v3.7.1/SLiM_Manual.pdf)!
:::
:::

::: {.column width="40%"}
<center>

![Modified from [Alexei Drummond](http://alexeidrummond.org/bayesian_phylo_lectures/lecture10/)](images/sim_sketches.001.png){width="100%"}

</center>
:::
:::

## Simple neutral simulation in SLiM

```{bash}
#| echo: false
cat examples/script.slim
```

## 

::: columns
::: {.column width="60%"}
<h2>What is _msprime_</h2>

::: fragment
-   A Python module for writing **coalescent simulations**
:::

::: fragment
-   Extremely fast (genome-scale, population-scale data)
:::

::: fragment
-   You must know Python fairly well to build complex models
:::
:::

::: {.column width="40%"}
<center>

![Modified from [Alexei Drummond](http://alexeidrummond.org/bayesian_phylo_lectures/lecture10/)](images/sim_sketches.002.png){width="100%"}

</center>
:::
:::

## Simple simulation using *msprime*

```{bash}
#| echo: false
cat examples/script.py
```

source: [link](https://tskit.dev/msprime/docs/stable/demography.html#demographic-models)

# 

<center>

<h2>[www.slendr.net](https://www.slendr.net)</h2>

![](images/slendr_logo.png){width="30%"}

</center>

::: fragment

<center>
<h2>Why a new package?</h2>
</center>

:::

## Spatial simulations!

<center>![](images/animation.gif){width="70%"}</center>

## Why a new package?

::: fragment
-   Most researchers are not expert programmers

-   All but the most trivial simulations require lots of code
:::

::: fragment
-   90% <sup><font color="blue">[citation needed]</font></sup> of simulations are basically the same!

    -   create populations (splits and $N_e$ changes)

    -   specify if/how they should mix (rates and times)
:::

::: fragment
-   Lot of code duplication across projects
:::

::: fragment
<center>***slendr*** **makes "standard" simulations trivial and<br> unlocks new kinds of spatial simulations**</center>
:::

# Let's get started

## Everything we do will be in R

<br>

Always start your R scripts with this (*):

```{r}
library(slendr)
init_env()
```

<br>

My solutions will also use these two packages:

```{r}
library(ggplot2)
library(dplyr)
```

::: aside
(*) You can safely ignore the message about missing SLiM.
:::


<!-- ## Workaround for an RStudio bug -->

<!-- RStudio sometimes interferes with Python setup that we need for simulation. To fix this, go to `Tools` -\> `Global Options` in your RStudio and set the following options: -->

<!-- <center>![](images/rstudio_setting.png){width="40%"}</center> -->


## Typical steps of a _slendr_ R workflow

<br>

1.  creating populations
2.  scheduling population splits
3.  programming $N_e$ size changes
4.  encoding gene-flow events
5.  simulation sequence of a given size
6.  computing statistics from simulated outputs


# 

::: {style="text-align: right"}
<h1>*slendr* haiku</h1>
:::

<br>

::: {style="text-align: right"}
Build simple models,

<br>

simulate data from them.

<br>

Just one plain R script.
:::

## Creating a `population()`

Each needs a name, size and time of appearance (i.e., "split"):

```{r}
pop1 <- population("pop1", N = 1000, time = 1)
```

. . .

<br>

This creates a normal R object. Typing it out gives a summary:

```{r}
pop1
```

## Programming population splits

Splits are indicated by the `parent = <pop>` argument:

```{r}
pop2 <- population("pop2", N = 100, time = 50, parent = pop1)
```

. . .

<br>

The split is reported in the "historical summary":

```{r}
pop2
```

## Scheduling resize events -- `resize()`

Step size decrease:

```{r}
#| code-line-numbers: "2"
pop1 <- population("pop1", N = 1000, time = 1)
pop1_step <- resize(pop1, N = 100, time = 500, how = "step")
```

<br>

Exponential increase:

```{r}
#| code-line-numbers: "2"
pop2 <- population("pop2", N = 100, time = 50, parent = pop1)
pop2_exp <- resize(pop2, N = 10000, time = 500, end = 2000, how = "exponential")
```

## Tidyverse-style [pipe](https://magrittr.tidyverse.org) `%>%` interface

A more concise way to express the same thing as before.

<br>

Step size decrease:

```{r}
pop1 <-
  population("pop1", N = 1000, time = 1) %>%
  resize(N = 100, time = 500, how = "step")
```

Exponential increase:

```{r}
pop2 <-
  population("pop2", N = 1000, time = 1) %>%
  resize(N = 10000, time = 500, end = 2000, how = "exponential")
```

## A more complex model

```{r}
pop1 <- population("pop1", N = 1000, time = 1)

pop2 <-
  population("pop2", N = 1000, time = 300, parent = pop1) %>%
  resize(N = 100, how = "step", time = 1000)

pop3 <-
  population("pop3", N = 1000, time = 400, parent = pop2) %>%
  resize(N = 2500, how = "step", time = 800)

pop4 <-
  population("pop4", N = 1500, time = 500, parent = pop3) %>%
  resize(N = 700, how = "exponential", time = 1200, end = 2000)

pop5 <-
  population("pop5", N = 100, time = 600, parent = pop4) %>%
  resize(N = 50, how = "step", time = 900) %>%
  resize(N = 1000, how = "exponential", time = 1600, end = 2200)
```

## Each object carries its history!

```{r}
pop5
```

## Last step before simulation: `compile_model()`

<br>

```{r}
model <- compile_model(
  list(pop1, pop2, pop3, pop4, pop5),
  generation_time = 1,
  simulation_length = 3000
)
```

<br>

<center>**Compilation takes a list of model components, performs internal consistency checks, returns a single model object.**</center>

## Model summary

Typing the compiled `model` into R prints a brief summary:

```{r}
model
```

## Model visualization

```{r}
#| fig-width: 7
#| fig-align: center
plot_model(model)
```


# Exercise #1

<br><br><br><br>

::: aside
Use [these](https://bodkan.quarto.pub/2023-08-23-ku-popgen23-onepage/) one-page handouts for reference.
:::

## Exercise #1: write this model in _slendr_

::: columns

::: {.column width="50%"}
![](images/intro_model1.png)
:::

::: {.column width="50%"}

::: {.fragment}
Start a `model1.R` script in [RStudio](http://emily.popgen.dk:3838/) with this 'template':

```{r}
#| eval: false
library(slendr)
init_env()

chimp <- population(...)
# <... rest of your code ...>

model <- compile_model(
  populations = list(chimp, ...),
  generation_time = 30
)

plot_model(model) # verify visually
```
:::
:::

:::

::: fragment
Don't worry about gene flow yet. We will add it at a later stage.
:::

# Exercise #1: solution

<br><br><br><br>

::: aside
[`ex1.R` script on GitHub](https://github.com/bodkan/ku-popgen2023/blob/main/solutions/ex1.R) one-page handouts for reference.
:::





#

<h1>So, now we can<br>write a `model`</h1>









## How do we simulate from it?

*slendr* has two built-in simulation "engine scripts":

- SLiM engine ([source](https://github.com/bodkan/slendr/blob/main/inst/scripts/script.slim))
- *msprime* engine ([source](https://github.com/bodkan/slendr/blob/main/inst/scripts/script.py))

They are designed to understand *slendr* models.

. . .

<br>All you need to simulate data is this one line of code:

```{r}
#| eval: false
ts <- msprime(model, sequence_length = 100e6, recombination_rate = 1e-8)
```

<center>

<br>

<h3>**You don't have to write *msprime* or SLiM code!**</h3>

</center>



# The output of a *slendr* simulation is a **tree sequence (`ts`)**

## What is tree sequence?

![](images/tree_sequence_diagram_no_muts.png){width="80%" fig-align="center"}

-   a record of full genetic ancestry of a set of samples
-   an encoding of DNA sequence carried by those samples
-   an efficient analysis framework

# Why tree sequence?

<br>

<h3>Why not VCF or a normal genotype table?</h3>

## What we usually have

<center>![](images/vcf_screenshot.png){width="90%"}</center>

## What we usually *want*

An understanding of our samples' evolutionary history:

<center>![](images/tree_sequence_diagram_no_muts.png)</center>

::: fragment
<center><h3>**This is exactly what a tree sequence *is*!**</h3></center>
:::

## The magic of tree sequences

They allow computing of popgen statistics *without genotypes*!

<center>

![](images/tree_sequence_diagram_no_muts.png)

</center>


There is a "duality" between mutations and branch lengths.

::: aside
See an amazing paper by [Ralph et al. (2020)](https://academic.oup.com/genetics/article/215/3/779/5930459) for more detail.
:::




## What if we need mutations though?

::: fragment
Coalescent and mutation processes can be decoupled!

<center>

![](images/tree_sequence_diagram_no_muts.png)

</center>

:::

## What if we need mutations though?


Coalescent and mutation processes can be decoupled!

<center>

![](images/tree_sequence_diagram.png)

::: fragment
<h3>With _slendr_, we can add mutations *after* the simulation using `ts_mutate()`.</h3>
:::

</center>






<!-- ## Tree-sequence simulation in practice -->

<!-- Let' say we have a compiled `model` object. -->

<!-- Then, we can simulate a tree sequence like this: -->

<!-- ```{r} -->
<!-- ts <- msprime(model, sequence_length = 10e6, recombination_rate = 1e-8) -->
<!-- ``` -->

<!-- But we can also simulate mutations like this: -->

<!-- ```{r} -->
<!-- ts <- -->
<!--   msprime(model, sequence_length = 10e6, recombination_rate = 1e-8) %>% -->
<!--   ts_mutate(mutation_rate = 1e-8) -->
<!-- ``` -->

<!-- <br> -->

<!-- <center>We will be using `ts_mutate()` throughout.</center> -->









<!-- ## Tree sequences are _very_ efficient -->

<!-- <br> -->

<!-- This simulates 2 $\times$ 10000 chromosomes of 100 Mb: -->

<!-- ```{r} -->
<!-- #| eval: false -->
<!-- pop <- population("pop", time = 1e6, N = 10000) -->
<!-- model <- compile_model(pop, generation_time = 30, direction = "backward") -->
<!-- ts <- msprime(model, sequence_length = 100e6, recombination_rate = 1e-8) -->
<!-- ``` -->

<!-- . . . -->

<!-- <br> -->

<!-- **Runs in less than 30 seconds on my laptop!** -->

<!-- **Takes about 66 Mb of memory!** -->

<!-- ## How does this work?! -->

<!-- . . . -->

<!-- <center> -->

<!-- ![](images/tables.jpeg) -->

<!-- <center> -->

<!-- ## Tree-sequence tables -->

<!-- ::: row -->
<!-- ::: columns -->
<!-- ::: {.column width="60%"} -->
<!-- A tree (sequence) can be represented by -->

<!-- ::: incremental -->
<!-- -   a table of <font color="orange">n</font><font color="green">o</font><font color="darkblue">d</font><font color="green">e</font><font color="darkblue">s</font>, -->
<!-- -   a table of [edges]{.underline} between nodes, -->
<!-- -   a table of <font color="red">mutations</font> on edges -->
<!-- ::: -->
<!-- ::: -->

<!-- ::: {.column width="40%"} -->
<!-- <br> -->

<!-- <center>![](images/tree_diagram.png)</center> -->
<!-- ::: -->
<!-- ::: -->
<!-- ::: -->

<!-- . . . -->

<!-- <center> -->

<!-- <h3>**A set of such tables is a tree sequence.**</h3> -->

<!-- </center> -->

<!-- ::: aside -->
<!-- You can find much more information in  ([tskit docs](https://tskit.dev/tutorials/tables_and_editing.html)) -->
<!-- ::: -->

<!-- ## Tree-sequence tables in practice -->

<!-- ::: columns -->
<!-- ::: {.column width="50%"} -->
<!-- ```{r} -->
<!-- #| echo: false -->
<!-- #| fig-height: 13 -->
<!-- set.seed(123) -->

<!-- ts <- msprime(model, sequence_length = 1e6, recombination_rate = 1e-8, random_seed = 42) %>% ts_mutate(1e-8, random_seed = 42) -->

<!-- # make a tiny example simplified tree sequence -->
<!-- ts_tiny <- ts_samples(ts) %>% sample_n(4) %>% pull(name) %>% ts_simplify(ts, simplify_to = .) -->
<!-- # extract tree #1 as an ape tree and also a tskit tree -->
<!-- t_phylo <- ts_phylo(ts_tiny, 1, quiet = TRUE) -->
<!-- t_tskit <- ts_tree(ts_tiny, 1) -->

<!-- # plot the phylo tree with ape -->
<!-- suppressPackageStartupMessages(library(ggtree)) -->
<!-- nodes <- ts_nodes(t_phylo) %>% as_tibble %>% dplyr::select(node = phylo_id, pop, node_id) -->
<!-- ggtree(t_phylo, branch.length = "none") %<+% nodes + -->
<!--   geom_label(aes(label = node_id), size = 15) +  -->
<!--   guides(color = "none") -->
<!-- ``` -->
<!-- ::: -->

<!-- ::: {.column width="50%"} -->
<!-- ::: fragment -->
<!-- nodes: -->

<!-- ```{r} -->
<!-- #| echo: false -->
<!-- ts_nodes(t_phylo) %>% head(3) %>% .[, c("node_id", "pop_id", "time")] %>% as.data.frame() -->
<!-- ``` -->
<!-- ::: -->

<!-- ::: fragment -->
<!-- edges: -->

<!-- ```{r} -->
<!-- #| echo: false -->
<!-- ts_edges(t_phylo) %>% head(3) %>% .[, c("child_node_id", "parent_node_id")] %>% as.data.frame() -->
<!-- ``` -->
<!-- ::: -->

<!-- ::: fragment -->
<!-- mutations: -->

<!-- ```{r} -->
<!-- #| echo: false -->
<!-- ts_table(ts_tiny, "mutations") %>% filter(node %in% c(53, 22, 20, 74, 9)) %>% head(3) %>% .[, c("id", "site", "node", "time")] %>% as.data.frame() -->
<!-- ``` -->
<!-- ::: -->
<!-- ::: -->
<!-- ::: -->





## Let's take the `model` from earlier...

```{r}
#| fig-align: center
#| echo: false
plot_model(model)
```

## ... and simulate data from it

<br>

In our script we'll have something like this:

```{r}
#| eval: false
#| code-line-numbers: 1-3|5-6
<... population() definitions ...>

model <- compile_model(...)
  
ts <-
  msprime(model, sequence_length = 50e6, recombination_rate = 1e-8)
```


## ... and simulate data from it

<br>

In our script we'll have something like this:

```{r}
#| eval: false
#| code-line-numbers: "7"
<... population() definitions ...>

model <- compile_model(...)
  
ts <-
  msprime(model, sequence_length = 50e6, recombination_rate = 1e-8) %>%
  ts_mutate(mutation_rate = 1e-8)
```


```{r}
#| echo: false
ts <-
  msprime(model, sequence_length = 50e6, recombination_rate = 1e-8) %>%
  ts_mutate(mutation_rate = 1e-8)
```

::: fragment
<br>

<center><h2>How do we work with this `ts` thing?</h2></center>
:::

<!-- ## Conversion to other genotype formats -->

<!-- If you have a tree-sequence object `ts`, you can do... -->

<!-- ```{r} -->
<!-- #| eval: false -->
<!-- ts_vcf(ts, path = "path/to/a/file.vcf.gz") -->
<!-- ``` -->

<!-- ```{r} -->
<!-- #| eval: false -->
<!-- ts_eigenstrat(ts, prefix = "path/to/eigenstrat/prefix") -->
<!-- ``` -->

<!-- ```{r} -->
<!-- #| echo: false -->
<!-- ts <- ts_simplify(ts, simplify_to = c("pop1_1", "pop2_1")) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- ts_genotypes(ts) -->
<!-- ``` -->


## *slendr*'s R interface to [*tskit*](https://tskit.dev/tskit) statistics

<center>

![](images/slendr_tskit.png)

</center>

Allele-frequecy spectrum, nucleotide diversity $\pi$, $F_{ST}$, etc.

## Extracting sample information

We can get samples recorded in `ts` with `ts_samples()`:

```{r}
#| eval: false
ts_samples(ts)
```

```{r}
#| echo: false
#| output-location: fragment
ts_samples(ts) %>% head(5)
```

. . .

Sometimes a shortcut `ts_names()` can be useful:

```{r}
#| eval: false
# gives a single vector of all names
ts_names(ts)
```

```{r}
#| echo: false
ts_names(ts) %>% head(5)
```

## Example: [allele frequency spectrum](https://en.wikipedia.org/wiki/Allele_frequency_spectrum)

```{r}
#| echo: false
pop <- population("pop", N = 10000, time = 1)

model <- compile_model(pop, generation_time = 1, simulation_length = 10000)

ts <-
  msprime(model, sequence_length = 10e6, recombination_rate = 1e-8) %>%
  ts_mutate(mutation_rate = 1e-8)
```


```{r}
#| echo: false
set.seed(42)
```

::: columns
::: {.column width="45%"}

Sample 5 individuals:

```{r}
names <- ts_names(ts)[1:5]
names
```

::: {.fragment}

<br>

Compute the AFS:

```{r}
afs <- ts_afs(ts, list(names))

afs[-1]
```
:::

:::

::: {.column width="2%"}
¬†
:::

::: {.column width="53%"}
::: fragment
```{r}
#| eval: false
plot(afs[-1], type = "b",
     xlab = "allele count bin",
     ylab = "frequency")
```

```{r}
#| echo: false
#| fig-height: 8
plot(afs[-1], type = "b",
     xlab = "allele count bin",
     ylab = "frequency", lwd = 3,
     cex = 1.5, cex.lab=1.5, cex.axis=1.5, cex.main=1.5, cex.sub=1.5)
```
:::
:::
:::

# Exercise #2


<br><br><br><br>

::: aside
Use [these](https://bodkan.quarto.pub/2023-08-23-ku-popgen23-onepage/) one-page handouts for reference.
:::


## Exercise #2: estimating $N_e$ from AFS

```{r}
#| echo: false
set.seed(42)
TRUE_NE <- 6543

pop <- population("pop", N = TRUE_NE, time = 100000)
model <- compile_model(pop, generation_time = 1, direction = "backward")

ts <-
  msprime(model, sequence_length = 10e6, recombination_rate = 1e-8, random_seed = 42) %>%
  ts_mutate(mutation_rate = 1e-8, random_seed = 42)

samples <- ts_names(ts) %>% sample(10)

afs_observed <- ts_afs(ts, list(samples))
```

Imagine you sequenced 10 samples from a population and computed this AFS vector (# singletons, doubletons, etc.):

<!-- ```{r} -->

<!-- #| echo: false -->

<!-- dput(as.vector(observed_afs)) -->

<!-- ``` -->

```{r}
afs_observed <- c(2520, 1449, 855, 622, 530, 446, 365, 334, 349, 244,
                  264, 218,  133, 173, 159, 142, 167, 129, 125, 143)
```

. . .

You know that the population had constant $N_e$ somewhere between 1000 and 30000 for the past 100,000 generations, and had mutation and recombination rates of 1e-8.

. . .

<hr>

**Try to guess the true ** $N_e$ given the observed AFS by running single-population simulations for a range of $N_e$ values and comparing `ts_afs()` results to `afs_observed`.

## Exercise #2: hints

1.  Write a function (in a script `model2.R`) with `Ne` as its only input. It will create one population, compile a model, run `ts_afs()` for 10 samples, and return the AFS vector.

::: fragment
2.  Find `Ne` that will give the closest AFS to the observed AFS:

    -   _option #1_ [_easy_]: Plot AFS for various $N_e$ values and eyeball which AFS looks closest judging from the plot

    -   _option #2_ [_harder_]: Simulate AFS vectors in steps of possible `Ne` (maybe `lapply()`?), find the [closest](https://en.wikipedia.org/wiki/Mean_squared_error) AFS
:::

# Exercise #2: solution

<br><br><br><br>

::: aside
[`ex2.R` script on GitHub](https://github.com/bodkan/ku-popgen2023/blob/main/solutions/ex2.R) one-page handouts for reference.
:::










# Gene flow / admixture

## We use the function `gene_flow()`

<br>

If we have populations `p1` and `p2`, we schedule gene flow as:

```{r}
#| eval: false
gf <- gene_flow(from = p1, to = p2, start = 500, end = 600, rate = 0.13)
```

. . .

<br>

Multiple gene-flow events can be gathered in a list:

```{r}
#| eval: false
gf <- list(
  gene_flow(from = p1, to = p2, start = 500, end = 600, rate = 0.13),
  gene_flow(from = ..., to = ..., start = ..., end = ..., rate = ...),
  ...
)
```

. . .

<br>

<center>**`gene_flow()` checks for consistency!**</center>


## Plugging gene flow(s) into a model

<br>
If you have a _slendr_ script, you can do this:
<br>

```{r}
#| eval: false
#| code-line-numbers: 3-7|9-13|11
<... population() definitions ...>

gf <- list(
  gene_flow(from = ... to = ..., start = ..., end = ..., rate = ...),
  gene_flow(from = ..., to = ..., start = ..., end = ..., rate = ...),
  ...
)

model <- compile_model(
  populations = list(...),
  gene_flow = gf,           # <--- gene flow(s) specified
  generation_time = <...>
)
```





# Exercise #3

## Exercise #3: adding gene flow

Add NEA ‚Üí EUR introgression to your `model1.R` script:

```{r}
#| echo: false
#| fig-align: center
chimp <- population("CHIMP", time = 7e6, N = 5000)
afr <- population("AFR", parent = chimp, time = 6e6, N = 15000)
eur <- population("EUR", parent = afr, time = 70e3, N = 3000)
nea <- population("NEA", parent = afr, time = 600e3, N = 1000, remove = 40e3)

model <- compile_model(populations = list(chimp, nea, afr, eur), generation_time = 30)

cowplot::plot_grid(
  plot_model(model) + ggtitle("Normal time scale"),
  plot_model(model, log = TRUE, proportions = TRUE) + ggtitle("Logarithmic time scale")
)
```

&nbsp;

## Exercise #3: adding gene flow

Add NEA ‚Üí EUR introgression to your `model1.R` script:

```{r}
#| echo: false
#| fig-align: center
gf <- gene_flow(from = nea, to = eur, rate = 0.03, start = 55000, end = 50000)

model <- compile_model(
  populations = list(chimp, nea, afr, eur),
  gene_flow = gf,
  generation_time = 30
)

cowplot::plot_grid(
  plot_model(model, proportions = TRUE) + ggtitle("Normal time scale"),
  plot_model(model, log = TRUE, proportions = TRUE) + ggtitle("Logarithmic time scale")
)
```

3% pulse (`rate = 0.03`) between 55-50 kya (`start`-`end`)


# Exercise #3: solution

<br><br><br><br>

::: aside
[`ex3_ex4.R` script on GitHub](https://github.com/bodkan/ku-popgen2023/blob/main/solutions/ex3.R) one-page handouts for reference.
:::












# Exercise #4

## Exercise #4: more statistics!

Practice computing tree seqence statistics on the `ts` object you simulated in your `model1.R` script.

```{r}
#| fig-align: center
#| fig-width: 7
#| echo: false
plot_model(model, proportions = TRUE, order = c("NEA", "EUR", "AFR", "CHIMP"), log = TRUE)
```

## Exercise #4: more statistics!

**Compute (some of) these on your tree sequence:**

- nucleotide [`ts_diversity()`](https://www.slendr.net/reference/ts_diversity.html#ref-examples) in each population <br>

- [`ts_divergence()`](https://www.slendr.net/reference/ts_divergence.html#ref-examples) between populations<br>

- [`ts_fst()`](https://www.slendr.net/reference/ts_fst.html#ref-examples) between populations<br>

- outgroup [`ts_f3()`](https://www.slendr.net/reference/ts_f4ratio.html#ref-examples) using CHIMP as the outgroup (A!) for different pairs of "B" and "C" populations<br>

- [`ts_f4()`](https://www.slendr.net/reference/ts_f4ratio.html#ref-examples) test of Neanderthal introgression in EUR

  \[hint: $f_4$(\<AFR\>, \<EUR\>; \<NEAND\>, \<CHIMP\>)\]

<center>**Do results match expectations given your model?**</center>


## Exercise #4: hint

For multipopulation statistics, you need a (named) list of samples in each population.

`ts_names()` has a `split = "pop"` option just for this:

```{r}
#| eval: false
samples <- ts_names(ts, split = "pop")
samples
```

```{r}
#| echo: false
ts_tmp <- msprime(model, sequence_length = 1, recombination_rate = 0)
tmp_list <- ts_names(ts_tmp, split = "pop") %>% lapply(`[`, 1:3) %>% .[c("AFR", "NEA")]
tmp_list
```

You can use this in place of `sample_names =` in code like:

```{r}
#| eval: false
ts_diversity(ts, sample_sets = samples)
```



# Exercise #4: solution

<br><br><br><br>

::: aside
[`ex3_ex4.R` script on GitHub](https://github.com/bodkan/ku-popgen2023/blob/main/solutions/ex4.R) one-page handouts for reference.
:::


## More information

<br>

- _slendr_ paper is now in [PCI EvolBiol](https://evolbiol.peercommunityin.org/articles/rec?id=630)

- documentation, tutorials is [here](https://www.slendr.net)

- GitHub repo (bug reports!) is [here](http://github.com/bodkan/slendr)

<hr>

- check out my new [_demografr_](http://github.com/bodkan/demografr) inference package 

<hr>

- contact details at [bodkan.net](https://bodkan.net)







# 


<h1><font color="orange">B</font><font color="darkblue">O</font><font color="green">N</font><font color="red">U</font><font color="pink">S</font> <font color="purple">H</font><font color="orange">O</font><font color="green">M</font><font color="lightblue">E</font><font color="cyan">W</font><font color="green">O</font><font color="red">R</font><font color="orange">K</font>





# Time-series data

## Sampling aDNA samples through time

<br>

Imagine we have `pop1`, `pop2`, ... compiled in a `model`.

<br>

To record *ancient* individuals in the tree sequence, we can use `schedule_sampling()` like this:

```{r}
#| eval: false
schedule_sampling(
  model,                # compiled slendr model object
  times = c(100, 500),  # at these times (can be also a single number) ...
  list(pop1, 42),       # ... sample 42 individuals from pop1
  list(pop2, 10),       # ... sample 10 individuals from pop2
  list(pop3, 1)         # ... sample 1 individual from pop 3
) 
```

## Sampling schedule format

The output of `schedule_sampling()` is a plain data frame:

```{r}
schedule_sampling(model, times = c(40000, 30000, 20000, 10000), list(eur, 1))
```

. . .

We can bind multiple sampling schedules together, giving us finer control about sampling:

```{r}
eur_samples <- schedule_sampling(model, times = c(40000, 30000, 20000, 10000, 0), list(eur, 1))
afr_samples <- schedule_sampling(model, times = 0, list(afr, 1))

samples <- bind_rows(eur_samples, afr_samples)
```

## How to use a sampling schedule?

To sample individuals based on a given schedule, we use the `samples =` argument of the `msprime()` functions:

```{r}
ts <- msprime(model, samples = samples, sequence_length = 1e6, recombination_rate = 1e-8) %>% 
  ts_mutate(mutation_rate = 1e-8)
```

. . .

We can verify that only specific individuals are recorded:

::: columns
::: {.column width="40%"}
```{r}
ts_samples(ts)
```
:::

::: {.column width="60%"}
::: fragment
```{r}
ts
```
:::
:::
:::

# Exercise #5

## Exercise #5a: ancient samples

```{r}
#| echo: false
chimp <- population("CHIMP", time = 7e6, N = 5000)
afr <- population("AFR", parent = chimp, time = 6e6, N = 15000)
eur <- population("EUR", parent = afr, time = 70e3, N = 3000)
nea <- population("NEA", parent = afr, time = 600e3, N = 1000, remove = 40e3)

gf <- gene_flow(from = nea, to = eur, rate = 0.03, start = 50000, end = 40000)

model <- compile_model(
  populations = list(chimp, nea, afr, eur),
  gene_flow = gf,
  generation_time = 30
)

plot_model(model)
```

## Exercise #5a: ancient samples

Simulate data from your model using this sampling:

-   one present-day CHIMP and AFR individual
-   20 present-day EUR individuals
-   1 NEA at 70 ky, 1 NEA at 40 ky
-   1 EUR every 1000 years between 50-5 kya

Reminder: you can do this by:

```{r}
#| eval: false
samples <- # rbind(...) together individual schedule_sampling() data frames

ts <-
  msprime(model, samples = samples, sequence_length = 100e6, recombination_rate = 1e-8) %>%
  ts_mutate(mutation_rate = 1e-8)
```

## Exercise #5b: $f_4$-ratio statistic

Use $f_4$-ratio statistic to replicate the [following figure](https://www.pnas.org/doi/10.1073/pnas.1814338116#fig01):

![](images/neand_decline.png){width="60%" fig-align="center"}

::: aside
Hint: You can compute Neanderthal ancestry for a vector of individuals `X` as `ts_f4ratio(ts, X = X, "NEA_1", "NEA_2", "AFR_1", "CHIMP_1")`.
:::

# Exercise #5: solution

<br><br><br><br>

::: aside
[`ex5.R` script on GitHub](https://github.com/bodkan/ku-popgen2023/blob/main/solutions/ex5.R) one-page handouts for reference.
:::
