<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Martin Petr">
<meta name="author" content="CC BY 4.0">

<title>Simulations in population genetics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="popgen2023_files/libs/clipboard/clipboard.min.js"></script>
<script src="popgen2023_files/libs/quarto-html/quarto.js"></script>
<script src="popgen2023_files/libs/quarto-html/popper.min.js"></script>
<script src="popgen2023_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="popgen2023_files/libs/quarto-html/anchor.min.js"></script>
<link href="popgen2023_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="popgen2023_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="popgen2023_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="popgen2023_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="popgen2023_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
</div>
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Simulations in population genetics</h1>
<p class="subtitle lead"><a href="http://www.popgen.dk/popgen23/">PopGen summer course 2023</a></p>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Authors</div>
    <div class="quarto-title-meta-contents">
             <p>Martin Petr </p>
             <p><a href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a> </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">August 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="first-things-first" class="level1">
<h1>First things first</h1>
<p>These slides and other resources are (and always will be) at:</p>
<p><a href="https://github.com/bodkan/ku-popgen2023">github.com/bodkan/ku-popgen2023</a></p>
<p><br><br></p>
<p><strong>Open this link now so you have everything at hand later.</strong></p>
</section>
<section id="first-things-first-1" class="level1">
<h1>First things first</h1>
<p>In order to make sure you can do the exercises, you should log into your <a href="emily.popgen.dk:3838/">online RStudio session</a> and run this in your R console:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">setup_env</span>(<span class="at">agree =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<p>If you want to do the exercises on your own laptop (if you <em>can</em>, you <em>should</em>), you need to do this first before running the above:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"slendr"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="section" class="level1">
<h1></h1>
<div class="columns">
<div class="column" style="width:70%;">
<blockquote class="blockquote">
<p>Many problems in population genetics cannot be solved by a mathematician, no matter how gifted. [It] is already clear that computer methods are very powerful. This is good. It […] <strong>permits people with limited mathematical knowledge to work on important problems</strong> […].</p>
</blockquote>
</div><div class="column" style="width:30%;">
<p><img src="images/crow.jpeg" class="img-fluid"></p>
<p><a href="https://en.wikipedia.org/wiki/James_F._Crow">James F. Crow</a> – <a href="http://www.gnxp.com/blog/2006/06/10-questions-for-jim-crow.php">interview</a></p>
</div>
</div>
</section>
<section id="why-use-simulations" class="level1 page-columns page-full">
<h1>Why use simulations?</h1>
<ol type="1">
<li>Making sense of inferred statistics</li>
<li>Fitting model parameters (i.e.&nbsp;<a href="https://en.wikipedia.org/wiki/Approximate_Bayesian_computation">ABC</a>)</li>
<li>Ground truth for method development</li>
</ol>
<section id="making-sense-of-inferred-statistics" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="making-sense-of-inferred-statistics">Making sense of inferred statistics</h2>
<center>
<img src="images/fstats_sims.png" class="img-fluid">
</center>

<div class="no-row-height column-margin column-container"><div class="">
<p>Image from <a href="https://academic.oup.com/genetics/article/202/4/1485/5930214">Peter (2016)</a></p>
</div></div></section>
<section id="fitting-model-parameters-i.e.-abc" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="fitting-model-parameters-i.e.-abc">Fitting model parameters (i.e.&nbsp;<a href="https://en.wikipedia.org/wiki/Approximate_Bayesian_computation">ABC</a>)</h2>
<center>
<img src="images/abc_scheme.png" class="img-fluid" style="width:50.0%">
</center>

<div class="no-row-height column-margin column-container"><div class="">
<p>Image from <a href="https://en.wikipedia.org/wiki/Approximate_Bayesian_computation">Wikipedia on ABC</a></p>
</div></div></section>
<section id="ground-truth-for-method-development" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="ground-truth-for-method-development">Ground truth for method development</h2>
<center>
<img src="images/mcmc.png" class="img-fluid">
</center>

<div class="no-row-height column-margin column-container"><div class="">
<p>Image from <a href="https://www.nature.com/articles/ng.3015">Schiffels and Durbin (2014)</a></p>
</div></div></section>
</section>
<section id="what-does-it-mean-to-simulate-a-genome" class="level1">
<h1>What does it mean to simulate a genome?</h1>
<p><br> How would you design an algorithm for a popgen simulation?</p>

<p><br>What minimum components are needed?</p>
</section>
<section id="if-we-want-to-simulate-population-genetics" class="level1">
<h1>If we want to simulate population genetics</h1>
<p><br></p>
<p>. . .</p>
<p>We need <em>populations</em>.</p>
<p>. . .</p>
<p>We need <em>genetics</em>.</p>
</section>
<section id="the-genetics-part" class="level1">
<h1>The ‘genetics’ part…</h1>
<p><br></p>
<p>. . .</p>
<p>…a linear sequence of nucleotides (a chromosome)</p>
<ul>
<li>a list of characters (A/G/C/T nucleotides)</li>
<li>a list of 0 or 1 values (ancestral/derived allele),</li>
</ul>
<p>. . .</p>
<p>which (maybe) mutates at a given <em>mutation rate</em>,<br></p>
<p>and (maybe) recombines at a certain <em>recombination rate</em>.</p>
</section>
<section id="the-population-part" class="level1">
<h1>The ‘population’ part…</h1>
<p>. . .</p>
<p><br></p>
<p>… a collection of <em>individuals</em> at a given point in time,<br><br> each carrying <em>chromosomes</em> inherited from its parents.</p>
</section>
<section id="section-1" class="level1">
<h1></h1>
<h1>
Single-locus<br>Wright-Fisher<br>simulation in R
</h1>
<p><br></p>
<h3 class="anchored">
“What is the expected trajectory of a neutral allele under the influence of genetic drift?”
</h3>
<section id="from-fernandos-lecture-on-monday" class="level2">
<h2 class="anchored" data-anchor-id="from-fernandos-lecture-on-monday">From Fernando’s lecture on Monday…</h2>
<p><br></p>
<center>
<img src="images/wf_trajectory.png" class="img-fluid" style="width:60.0%">
</center>
</section>
<section id="a-complete-wright-fisher-simulation" class="level2">
<h2 class="anchored" data-anchor-id="a-complete-wright-fisher-simulation">A complete Wright-Fisher simulation</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>p_0 <span class="ot">&lt;-</span> <span class="fl">0.5</span>  <span class="co"># initial allele frequency ("50% marbles" from lecture #1)</span></span>
<span id="cb3-2"><a href="#cb3-2"></a>N   <span class="ot">&lt;-</span> <span class="dv">500</span>  <span class="co"># number of chromosomes in a population</span></span>
<span id="cb3-3"><a href="#cb3-3"></a>T   <span class="ot">&lt;-</span> <span class="dv">500</span>  <span class="co"># number of generations to simulate</span></span>
<span id="cb3-4"><a href="#cb3-4"></a></span>
<span id="cb3-5"><a href="#cb3-5"></a><span class="co"># a vector for storing frequencies over time</span></span>
<span id="cb3-6"><a href="#cb3-6"></a>p_trajectory <span class="ot">&lt;-</span> p_0 </span>
<span id="cb3-7"><a href="#cb3-7"></a></span>
<span id="cb3-8"><a href="#cb3-8"></a><span class="co"># in each generation:</span></span>
<span id="cb3-9"><a href="#cb3-9"></a><span class="cf">for</span> (gen_i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>T) {</span>
<span id="cb3-10"><a href="#cb3-10"></a>  <span class="co"># get frequency in the previous generation</span></span>
<span id="cb3-11"><a href="#cb3-11"></a>  p_prev <span class="ot">&lt;-</span> p_trajectory[gen_i <span class="sc">-</span> <span class="dv">1</span>] </span>
<span id="cb3-12"><a href="#cb3-12"></a></span>
<span id="cb3-13"><a href="#cb3-13"></a>  <span class="co"># calculate new frequency ("sampling marbles from a jar")</span></span>
<span id="cb3-14"><a href="#cb3-14"></a>  p_trajectory[gen_i] <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="dv">1</span>, N, p_prev) <span class="sc">/</span> N </span>
<span id="cb3-15"><a href="#cb3-15"></a>}</span>
<span id="cb3-16"><a href="#cb3-16"></a></span>
<span id="cb3-17"><a href="#cb3-17"></a><span class="fu">plot</span>(p_trajectory)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="n-500-p_0-0.5" class="level2">
<h2 class="anchored" data-anchor-id="n-500-p_0-0.5"><span class="math inline">\(N\)</span> = 500, <span class="math inline">\(p_0 = 0.5\)</span></h2>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(p_trajectory, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>),</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h =</span> p_0, <span class="at">lty =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">"red"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="popgen2023_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="lets-make-it-a-function" class="level2">
<h2 class="anchored" data-anchor-id="lets-make-it-a-function">Let’s make it a function</h2>
<p><strong>Input:</strong> <span class="math inline">\(p_0\)</span>, <span class="math inline">\(N\)</span>, and the number of generations</p>
<p><strong>Output:</strong> allele frequency trajectory vector</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>simulate <span class="ot">&lt;-</span> <span class="cf">function</span>(p_0, N, T) {</span>
<span id="cb5-2"><a href="#cb5-2"></a>  p_trajectory <span class="ot">&lt;-</span> p_0</span>
<span id="cb5-3"><a href="#cb5-3"></a></span>
<span id="cb5-4"><a href="#cb5-4"></a>  <span class="cf">for</span> (gen_i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>T) {</span>
<span id="cb5-5"><a href="#cb5-5"></a>    p_prev <span class="ot">&lt;-</span> p_trajectory[gen_i <span class="sc">-</span> <span class="dv">1</span>]</span>
<span id="cb5-6"><a href="#cb5-6"></a>    p <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="dv">1</span>, N, p_prev) <span class="sc">/</span> N</span>
<span id="cb5-7"><a href="#cb5-7"></a>    p_trajectory[gen_i] <span class="ot">&lt;-</span> p</span>
<span id="cb5-8"><a href="#cb5-8"></a>  }</span>
<span id="cb5-9"><a href="#cb5-9"></a></span>
<span id="cb5-10"><a href="#cb5-10"></a>  p_trajectory</span>
<span id="cb5-11"><a href="#cb5-11"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="n-500-p_0-0.5-20-replicates" class="level2">
<h2 class="anchored" data-anchor-id="n-500-p_0-0.5-20-replicates"><span class="math inline">\(N\)</span> = 500, <span class="math inline">\(p_0 = 0.5\)</span> (20 replicates)</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>reps <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">20</span>, <span class="fu">simulate</span>(<span class="at">N =</span> <span class="dv">500</span>, <span class="at">p_0 =</span> <span class="fl">0.5</span>, <span class="at">T =</span> <span class="dv">2000</span>))</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">matplot</span>(reps, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">lty =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2023_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="n-1000-p_0-0.5-20-replicates" class="level2">
<h2 class="anchored" data-anchor-id="n-1000-p_0-0.5-20-replicates"><span class="math inline">\(N\)</span> = 1000, <span class="math inline">\(p_0 = 0.5\)</span> (20 replicates)</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>reps <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">20</span>, <span class="fu">simulate</span>(<span class="at">N =</span> <span class="dv">1000</span>, <span class="at">p_0 =</span> <span class="fl">0.5</span>, <span class="at">T =</span> <span class="dv">2000</span>))</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">matplot</span>(reps, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">lty =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2023_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="n-5000-p_0-0.5-20-replicates" class="level2">
<h2 class="anchored" data-anchor-id="n-5000-p_0-0.5-20-replicates"><span class="math inline">\(N\)</span> = 5000, <span class="math inline">\(p_0 = 0.5\)</span> (20 replicates)</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>reps <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">20</span>, <span class="fu">simulate</span>(<span class="at">N =</span> <span class="dv">5000</span>, <span class="at">p_0 =</span> <span class="fl">0.5</span>, <span class="at">T =</span> <span class="dv">2000</span>))</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">matplot</span>(reps, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">lty =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2023_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="n-10000-p_0-0.5-20-replicates" class="level2">
<h2 class="anchored" data-anchor-id="n-10000-p_0-0.5-20-replicates"><span class="math inline">\(N\)</span> = 10000, <span class="math inline">\(p_0 = 0.5\)</span> (20 replicates)</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>reps <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">20</span>, <span class="fu">simulate</span>(<span class="at">N =</span> <span class="dv">10000</span>, <span class="at">p_0 =</span> <span class="fl">0.5</span>, <span class="at">T =</span> <span class="dv">2000</span>))</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="fu">matplot</span>(reps, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">lty =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2023_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="n-10000-p_0-0.5-100-replicates" class="level2">
<h2 class="anchored" data-anchor-id="n-10000-p_0-0.5-100-replicates"><span class="math inline">\(N\)</span> = 10000, <span class="math inline">\(p_0 = 0.5\)</span> (100 replicates)</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>reps <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">100</span>, <span class="fu">simulate</span>(<span class="at">N =</span> <span class="dv">10000</span>, <span class="at">p_0 =</span> <span class="fl">0.5</span>, <span class="at">T =</span> <span class="dv">30000</span>))</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="fu">matplot</span>(reps, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">lty =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2023_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="n-10000-p_0-0.5-100-replicates-1" class="level2">
<h2 class="anchored" data-anchor-id="n-10000-p_0-0.5-100-replicates-1"><span class="math inline">\(N\)</span> = 10000, <span class="math inline">\(p_0 = 0.5\)</span> (100 replicates)</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>factors <span class="ot">&lt;-</span> MASS<span class="sc">::</span><span class="fu">fractions</span>(<span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">10</span>))</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">matplot</span>(reps, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">lty =</span> <span class="dv">1</span>)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> <span class="dv">10000</span> <span class="sc">*</span> factors, <span class="at">lwd =</span> <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2023_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="expected-allele-frequency-distribution" class="level2">
<h2 class="anchored" data-anchor-id="expected-allele-frequency-distribution">Expected allele frequency distribution</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'dplyr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    filter, lag</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union</code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(parallel)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="st">"diffusion.rds"</span>)) {</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>p_0 <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>N <span class="ot">&lt;-</span> <span class="dv">10000</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>factors <span class="ot">&lt;-</span> MASS<span class="sc">::</span><span class="fu">fractions</span>(<span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">10</span>))</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>final_frequencies <span class="ot">&lt;-</span> parallel<span class="sc">::</span><span class="fu">mclapply</span>(</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">seq_along</span>(factors), <span class="cf">function</span>(i) {</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>    f <span class="ot">&lt;-</span> factors[i]</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>    t <span class="ot">&lt;-</span> <span class="fu">as.integer</span>(N <span class="sc">*</span> f)</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>    <span class="co"># get complete trajectories as a matrix (each column a single trajectory)</span></span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>    reps <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">10000</span>, <span class="fu">simulate</span>(<span class="at">N =</span> N, <span class="at">p_0 =</span> p_0, <span class="at">T =</span> t))</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>    <span class="co"># only keep the last slice of the matrix with the final frequencies</span></span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>(</span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>      <span class="at">t =</span> <span class="fu">sprintf</span>(<span class="st">"t = %s * N"</span>, f),</span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>      <span class="at">freq =</span> reps[t, ]</span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>  },</span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a>  <span class="at">mc.cores =</span> parallel<span class="sc">::</span><span class="fu">detectCores</span>()</span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span> <span class="fu">do.call</span>(rbind, .)</span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a>final_frequencies<span class="sc">$</span>t <span class="ot">&lt;-</span> forcats<span class="sc">::</span><span class="fu">fct_rev</span>(forcats<span class="sc">::</span><span class="fu">fct_relevel</span>(final_frequencies<span class="sc">$</span>t, <span class="fu">sprintf</span>(<span class="st">"t = %s * N"</span>, factors)))</span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(final_frequencies, <span class="st">"diffusion.rds"</span>)</span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb16-25"><a href="#cb16-25" aria-hidden="true" tabindex="-1"></a>  final_frequencies <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"diffusion.rds"</span>)</span>
<span id="cb16-26"><a href="#cb16-26" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb16-27"><a href="#cb16-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-28"><a href="#cb16-28" aria-hidden="true" tabindex="-1"></a>final_frequencies <span class="sc">%&gt;%</span> .[.<span class="sc">$</span>freq <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;</span> .<span class="sc">$</span>freq <span class="sc">&lt;</span> <span class="dv">1</span>, ] <span class="sc">%&gt;%</span></span>
<span id="cb16-29"><a href="#cb16-29" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb16-30"><a href="#cb16-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(freq, <span class="at">y =</span> <span class="fu">after_stat</span>(density), <span class="at">fill =</span> t), <span class="at">position =</span> <span class="st">"identity"</span>, <span class="at">bins =</span> <span class="dv">100</span>, <span class="at">alpha =</span> <span class="fl">0.75</span>) <span class="sc">+</span></span>
<span id="cb16-31"><a href="#cb16-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"allele frequency"</span>) <span class="sc">+</span></span>
<span id="cb16-32"><a href="#cb16-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">3</span>)) <span class="sc">+</span></span>
<span id="cb16-33"><a href="#cb16-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(t <span class="sc">~</span> .) <span class="sc">+</span></span>
<span id="cb16-34"><a href="#cb16-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">fill =</span> <span class="fu">guide_legend</span>(<span class="fu">sprintf</span>(<span class="st">"time since</span><span class="sc">\n</span><span class="st">the start</span><span class="sc">\n</span><span class="st">[assuming</span><span class="sc">\n</span><span class="st">N = %s]"</span>, N))) <span class="sc">+</span></span>
<span id="cb16-35"><a href="#cb16-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb16-36"><a href="#cb16-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">strip.text.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb16-37"><a href="#cb16-37" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>),</span>
<span id="cb16-38"><a href="#cb16-38" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>),</span>
<span id="cb16-39"><a href="#cb16-39" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>),</span>
<span id="cb16-40"><a href="#cb16-40" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>),</span>
<span id="cb16-41"><a href="#cb16-41" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>),</span>
<span id="cb16-42"><a href="#cb16-42" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2023_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="fragment">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/kimura.jpeg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption"><em>“Diffusion Models in Population Genetics”</em>, <a href="https://www.jstor.org/stable/3211856#metadata_info_tab_contents">Kimura (1964)</a></figcaption>
</figure>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="section-2" class="level1" data-background-image="images/montypython.jpeg">
<h1 data-background-image="images/montypython.jpeg"></h1>
<h1>
<br><br><br>
</h1>
<h1 color="black" style="background-color: white">
<p>&nbsp;&nbsp;But now for something<br>&nbsp;&nbsp;completely different.</p>
</h1>
</section>
<section id="section-3" class="level1" data-background-image="images/montypython.jpeg">
<h1 data-background-image="images/montypython.jpeg"></h1>
<h1>
<br><br><br>
</h1>
<h1 color="black" style="background-color: white">
<p>&nbsp;&nbsp;Let’s talk about<br>&nbsp;&nbsp;<em>real</em> simulations.</p>
</h1>
<section id="there-are-many-simulation-tools" class="level2">
<h2 class="anchored" data-anchor-id="there-are-many-simulation-tools">There are many simulation tools</h2>
<p>The most famous and widely used are <a href="https://messerlab.org/slim/">SLiM</a> and <a href="https://tskit.dev/msprime/docs/stable/intro.html"><em>msprime</em></a>.</p>
<p>They are <u>very</u> powerful but both require:</p>
<div class="fragment">
<ul>
<li>quite a bit of programming knowledge,</li>
<li>a lot of code for non-trivial simulations (🐛🪲🐜).</li>
</ul>
</div>
<div class="fragment">
<p><br></p>
<center>
<h3 class="anchored">
<strong>The exercises will focus on the <a href="http://www.slendr.net"><em>slendr</em></a></strong><br>popgen simulation toolkit for R.
</h3>
</center>
</div>
<div class="fragment">
<br>
<center>
But let’s look at SLiM and <em>msprime</em> at least a little bit.
</center>
</div>
</section>
<section id="section-4" class="level2">
<h2 class="anchored" data-anchor-id="section-4"></h2>
<div class="columns">
<div class="column" style="width:60%;">
<h2 class="anchored">
What is SLiM?
</h2>
<ul>
<li><strong>A forward-time simulator</strong></li>
</ul>
<div class="fragment">
<ul>
<li>It’s fully programmable!</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Massive library of functions for:
<ul>
<li>demographic events</li>
<li>various mating systems</li>
<li>selection, quantitative traits, …</li>
</ul></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>&gt; 700 pages long <a href="https://github.com/MesserLab/SLiM/releases/download/v3.7.1/SLiM_Manual.pdf">manual</a>!</li>
</ul>
</div>
</div><div class="column" style="width:40%;">
<center>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/sim_sketches.001.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption class="figure-caption">Modified from <a href="http://alexeidrummond.org/bayesian_phylo_lectures/lecture10/">Alexei Drummond</a></figcaption>
</figure>
</div>
</center>
</div>
</div>
</section>
<section id="simple-neutral-simulation-in-slim" class="level2">
<h2 class="anchored" data-anchor-id="simple-neutral-simulation-in-slim">Simple neutral simulation in SLiM</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>initialize() {
    // create a neutral mutation type
    initializeMutationType("m1", 0.5, "f", 0.0);
    // initialize 1Mb segment
    initializeGenomicElementType("g1", m1, 1.0);
    initializeGenomicElement(g1, 0, 999999);
    // set mutation rate and recombination rate of the segment
    initializeMutationRate(1e-8);
    initializeRecombinationRate(1e-8);
}

// create an ancestral population p1 of 10000 diploid individuals
1 early() { sim.addSubpop("p1", 10000); }

// in generation 1000, create two daughter populations p2 and p3
1000 early() {
    sim.addSubpopSplit("p2", 5000, p1);
    sim.addSubpopSplit("p3", 1000, p1);
}

// in generation 10000, stop the simulation and save 100 individuals
// from p2 and p3 to a VCF file
10000 late() {
    p2_subset = sample(p2.individuals, 100);
    p3_subset = sample(p3.individuals, 100);
    c(p2_subset, p3_subset).genomes.outputVCF("/tmp/slim_output.vcf.gz");
    sim.simulationFinished();
    catn("DONE!");
}</code></pre>
</div>
</div>
</section>
<section id="section-5" class="level2">
<h2 class="anchored" data-anchor-id="section-5"></h2>
<div class="columns">
<div class="column" style="width:60%;">
<h2 class="anchored">
What is <em>msprime</em>?
</h2>
<ul>
<li>A Python module for writing <strong>coalescent simulations</strong></li>
</ul>
<div class="fragment">
<ul>
<li>Extremely fast (genome-scale, population-scale data)</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>You must know Python fairly well to build complex models</li>
</ul>
</div>
</div><div class="column" style="width:40%;">
<center>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/sim_sketches.002.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption class="figure-caption">Modified from <a href="http://alexeidrummond.org/bayesian_phylo_lectures/lecture10/">Alexei Drummond</a></figcaption>
</figure>
</div>
</center>
</div>
</div>
</section>
<section id="simple-simulation-using-msprime" class="level2">
<h2 class="anchored" data-anchor-id="simple-simulation-using-msprime">Simple simulation using <em>msprime</em></h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>import msprime

demography = msprime.Demography()
demography.add_population(name="A", initial_size=10_000)
demography.add_population(name="B", initial_size=5_000)
demography.add_population(name="C", initial_size=1_000)
demography.add_population_split(time=1000, derived=["A", "B"], ancestral="C")

ts = msprime.sim_ancestry(
  sequence_length=10e6,
  recombination_rate=1e-8,
  samples={"A": 100, "B": 100},
  demography=demography
)</code></pre>
</div>
</div>
</section>
</section>
<section id="section-6" class="level1">
<h1></h1>
<center>
<h2 class="anchored" data-anchor-id="section-6">
<a href="https://www.slendr.net">www.slendr.net</a>
</h2>
<p><img src="images/slendr_logo.png" class="img-fluid" style="width:30.0%"></p>
</center>
<div class="fragment">
<center>
<h2 class="anchored">
Why a new package?
</h2>
</center>
</div>
<section id="spatial-simulations" class="level2">
<h2 class="anchored" data-anchor-id="spatial-simulations">Spatial simulations!</h2>
<center>
<img src="images/animation.gif" class="img-fluid" style="width:70.0%">
</center>
</section>
<section id="section-7" class="level2">
<h2 class="anchored" data-anchor-id="section-7"></h2>
<h2 class="anchored">
Why a new package?
</h2>
<ul>
<li><p>Most researchers are not expert programmers</p></li>
<li><p>All but the most trivial simulations require lots of code</p></li>
</ul>
<div class="fragment">
<ul>
<li><p>90% <sup><font color="blue">[citation needed]</font></sup> of simulations are basically the same!</p>
<ul>
<li><p>create populations (splits and <span class="math inline">\(N_e\)</span> changes)</p></li>
<li><p>specify if/how they should mix (rates and times)</p></li>
</ul></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Lot of code duplication across projects</li>
</ul>
</div>
<div class="fragment">
<center>
<strong><em>slendr</em></strong> <strong>makes “standard” demographic simulations trivial<br>and unlocks new kinds of spatial simulations</strong>
</center>
</div>
</section>
</section>
<section id="lets-get-started" class="level1 page-columns page-full">
<h1>Let’s get started</h1>
<section id="everything-we-do-will-be-in-r" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="everything-we-do-will-be-in-r">Everything we do will be in R</h2>
<p><br></p>
<p>Always start your R scripts with this (*):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,
which was just loaded, will retire in October 2023.
Please refer to R-spatial evolution reports for details, especially
https://r-spatial.org/r/2023/05/15/evolution4.html.
It may be desirable to make the sf package available;
package maintainers should consider adding sf to Suggests:.
The sp package is now running under evolution status 2
     (status 2 uses the sf package in place of rgdal)</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">init_env</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>The interface to all required Python modules has been activated.</code></pre>
</div>
</div>
<p><br></p>
<p>My solutions will also use these two packages:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<div class="no-row-height column-margin column-container"><div class="">
<p>(*) You can safely ignore the message about missing SLiM.</p>
</div></div></section>
<section id="typical-steps-of-a-slendr-r-workflow" class="level2">
<h2 class="anchored" data-anchor-id="typical-steps-of-a-slendr-r-workflow">Typical steps of a <em>slendr</em> R workflow</h2>
<p><br></p>
<ol type="1">
<li>creating populations</li>
<li>scheduling population splits</li>
<li>programming <span class="math inline">\(N_e\)</span> size changes</li>
<li>encoding gene-flow events</li>
<li>simulation sequence of a given size</li>
<li>computing statistics from simulated outputs</li>
</ol>
</section>
<section id="creating-a-population" class="level2">
<h2 class="anchored" data-anchor-id="creating-a-population">Creating a <code>population()</code></h2>
<p>Each needs a name, size and time of appearance (i.e., “split”):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>. . .</p>
<p><br></p>
<p>This creates a normal R object. Typing it out gives a summary:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>pop1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'population' object 
-------------------------- 
name: pop1 
non-spatial population
stays until the end of the simulation

population history overview:
  - time 1: created as an ancestral population (N = 1000)</code></pre>
</div>
</div>
</section>
<section id="programming-population-splits" class="level2">
<h2 class="anchored" data-anchor-id="programming-population-splits">Programming population splits</h2>
<p>Splits are indicated by the <code>parent = &lt;pop&gt;</code> argument:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>pop2 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">50</span>, <span class="at">parent =</span> pop1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>. . .</p>
<p><br></p>
<p>The split is reported in the “historical summary”:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>pop2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'population' object 
-------------------------- 
name: pop2 
non-spatial population
stays until the end of the simulation

population history overview:
  - time 50: split from pop1 (N = 100)</code></pre>
</div>
</div>
</section>
<section id="scheduling-resize-events-resize" class="level2">
<h2 class="anchored" data-anchor-id="scheduling-resize-events-resize">Scheduling resize events – <code>resize()</code></h2>
<p>Step size decrease:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span>
<span id="cb30-2"><a href="#cb30-2"></a>pop1_step <span class="ot">&lt;-</span> <span class="fu">resize</span>(pop1, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">how =</span> <span class="st">"step"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<p>Exponential increase:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a>pop2 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">50</span>, <span class="at">parent =</span> pop1)</span>
<span id="cb31-2"><a href="#cb31-2"></a>pop2_exp <span class="ot">&lt;-</span> <span class="fu">resize</span>(pop2, <span class="at">N =</span> <span class="dv">10000</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">2000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="tidyverse-style-pipe-interface" class="level2">
<h2 class="anchored" data-anchor-id="tidyverse-style-pipe-interface">Tidyverse-style <a href="https://magrittr.tidyverse.org">pipe</a> <code>%&gt;%</code> interface</h2>
<p>A more concise way to express the same thing as before.</p>
<p><br></p>
<p>Step size decrease:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>pop1 <span class="ot">&lt;-</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">how =</span> <span class="st">"step"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<p>Exponential increase:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>pop2 <span class="ot">&lt;-</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">10000</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">2000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="a-more-complex-model" class="level2">
<h2 class="anchored" data-anchor-id="a-more-complex-model">A more complex model</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>pop2 <span class="ot">&lt;-</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">300</span>, <span class="at">parent =</span> pop1) <span class="sc">%&gt;%</span></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">100</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">1000</span>)</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>pop3 <span class="ot">&lt;-</span></span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop3"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">400</span>, <span class="at">parent =</span> pop2) <span class="sc">%&gt;%</span></span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">2500</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">800</span>)</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>pop4 <span class="ot">&lt;-</span></span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop4"</span>, <span class="at">N =</span> <span class="dv">1500</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">parent =</span> pop3) <span class="sc">%&gt;%</span></span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">700</span>, <span class="at">how =</span> <span class="st">"exponential"</span>, <span class="at">time =</span> <span class="dv">1200</span>, <span class="at">end =</span> <span class="dv">2000</span>)</span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a>pop5 <span class="ot">&lt;-</span></span>
<span id="cb34-16"><a href="#cb34-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop5"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">600</span>, <span class="at">parent =</span> pop4) <span class="sc">%&gt;%</span></span>
<span id="cb34-17"><a href="#cb34-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">50</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">900</span>) <span class="sc">%&gt;%</span></span>
<span id="cb34-18"><a href="#cb34-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">1000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>, <span class="at">time =</span> <span class="dv">1600</span>, <span class="at">end =</span> <span class="dv">2200</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="each-object-carries-its-history" class="level2">
<h2 class="anchored" data-anchor-id="each-object-carries-its-history">Each object carries its history!</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>pop5</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'population' object 
-------------------------- 
name: pop5 
non-spatial population
stays until the end of the simulation

population history overview:
  - time 600: split from pop4 (N = 100)
  - time 900: resize from 100 to 50 individuals
  - time 1600-2200: exponential resize from 50 to 1000 individuals</code></pre>
</div>
</div>
</section>
<section id="last-step-before-simulation-compile_model" class="level2">
<h2 class="anchored" data-anchor-id="last-step-before-simulation-compile_model">Last step before simulation: <code>compile_model()</code></h2>
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(pop1, pop2, pop3, pop4, pop5),</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">generation_time =</span> <span class="dv">1</span>,</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">simulation_length =</span> <span class="dv">3000</span></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<center>
<strong>Compilation takes a list of model components, performs internal consistency checks, returns a single model object.</strong>
</center>
</section>
<section id="model-summary" class="level2">
<h2 class="anchored" data-anchor-id="model-summary">Model summary</h2>
<p>Typing the compiled <code>model</code> into R prints a brief summary:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>model</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'model' object 
--------------------- 
populations: pop1, pop2, pop3, pop4, pop5 
geneflow events: [no geneflow]
generation time: 1 
time direction: forward 
total running length: 3000 model time units
model type: non-spatial

configuration files in: /private/var/folders/d_/hblb15pd3b94rg0v35920wd80000gn/T/RtmprrhY6k/filee23c6b232dc7 </code></pre>
</div>
</div>
</section>
<section id="model-visualization" class="level2">
<h2 class="anchored" data-anchor-id="model-visualization">Model visualization</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2023_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="exercise-1" class="level1 page-columns page-full">
<h1>Exercise #1</h1>
<p><br><br><br><br></p>

<div class="no-row-height column-margin column-container"><div class="">
<p>Use <a href="https://bodkan.quarto.pub/ku-popgen2023-onepage/">these</a> one-page handouts for reference.</p>
</div></div><section id="exercise-1-write-this-model-in-slendr" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="exercise-1-write-this-model-in-slendr">Exercise #1: write this model in <em>slendr</em></h2>
<div class="columns">
<div class="column" style="width:45%;">
<p><img src="images/intro_model1.png" class="img-fluid"></p>
</div><div class="column" style="width:55%;">
<div class="fragment">
<p>Start a <code>model1.R</code> script in <a href="http://emily.popgen.dk:3838/">RStudio</a> with this “template”:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr)</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="fu">init_env</span>()</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>chimp <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"CHIMP"</span>, <span class="at">time =</span> <span class="fl">7e6</span>, <span class="at">N =</span> <span class="dv">5000</span>)</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a><span class="co"># &lt;... rest of your code ...&gt;</span></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">populations =</span> <span class="fu">list</span>(chimp, ...),</span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">generation_time =</span> <span class="dv">30</span></span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model) <span class="co"># verify visually</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>

<div class="no-row-height column-margin column-container"><div class="">
<p><strong>Note:</strong> With <em>slendr</em>, you can specify events in more natural “units of years ago”. Just program times in a “decreasing order” (7Mya → 6Mya → …, as shown in the picture above).</p>
</div></div></section>
</section>
<section id="exercise-1-solution" class="level1 page-columns page-full">
<h1>Exercise #1: solution</h1>
<p><br><br><br><br></p>

<div class="no-row-height column-margin column-container"><div class="">
<p>See <a href="https://github.com/bodkan/ku-popgen2023/blob/main/solutions/ex1.R"><code>ex1.R</code> on GitHub</a> for a solution.</p>
</div></div></section>
<section id="section-8" class="level1">
<h1></h1>
<h1>
So, now we can<br>write a <code>model</code>
</h1>
<section id="how-do-we-simulate-from-it" class="level2">
<h2 class="anchored" data-anchor-id="how-do-we-simulate-from-it">How do we simulate from it?</h2>
<p><em>slendr</em> has two built-in simulation “engine scripts”:</p>
<ul>
<li>SLiM engine (<a href="https://github.com/bodkan/slendr/blob/main/inst/scripts/script.slim">source</a>)</li>
<li><em>msprime</em> engine (<a href="https://github.com/bodkan/slendr/blob/main/inst/scripts/script.py">source</a>)</li>
</ul>
<p>They are designed to understand <em>slendr</em> models.</p>
<p>. . .</p>
<p><br>All you need to simulate data is this one line of code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<center>
<p><br></p>
<h3 class="anchored">
<strong>You don’t have to write <em>msprime</em> or SLiM code!</strong>
</h3>
</center>
</section>
</section>
<section id="the-output-of-a-slendr-simulation-is-a-tree-sequence-ts" class="level1">
<h1>The output of a <em>slendr</em> simulation is a <strong>tree sequence (<code>ts</code>)</strong></h1>
<section id="what-is-tree-sequence" class="level2">
<h2 class="anchored" data-anchor-id="what-is-tree-sequence">What is tree sequence?</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/tree_sequence_diagram_no_muts.png" class="img-fluid figure-img" style="width:80.0%"></p>
</figure>
</div>
<ul>
<li>a record of full genetic ancestry of a set of samples</li>
<li>an encoding of DNA sequence carried by those samples</li>
<li>an efficient analysis framework</li>
</ul>
</section>
</section>
<section id="why-tree-sequence" class="level1 page-columns page-full">
<h1>Why tree sequence?</h1>
<p><br></p>
<h3 class="anchored">
Why not VCF or a normal genotype table?
</h3>
<section id="what-we-usually-have" class="level2">
<h2 class="anchored" data-anchor-id="what-we-usually-have">What we usually have</h2>
<center>
<img src="images/vcf_screenshot.png" class="img-fluid" style="width:90.0%">
</center>
</section>
<section id="what-we-usually-want" class="level2">
<h2 class="anchored" data-anchor-id="what-we-usually-want">What we usually <em>want</em></h2>
<p>An understanding of our samples’ evolutionary history:</p>
<center>
<img src="images/tree_sequence_diagram_no_muts.png" class="img-fluid">
</center>
<div class="fragment">
<center>
<h3 class="anchored">
<strong>This is exactly what a tree sequence <em>is</em>!</strong>
</h3>
</center>
</div>
</section>
<section id="the-magic-of-tree-sequences" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="the-magic-of-tree-sequences">The magic of tree sequences</h2>
<p>They allow computing of popgen statistics <em>without genotypes</em>!</p>
<center>
<p><img src="images/tree_sequence_diagram_no_muts.png" class="img-fluid"></p>
</center>
<p>There is a “duality” between mutations and branch lengths.</p>

<div class="no-row-height column-margin column-container"><div class="">
<p>See an amazing paper by <a href="https://academic.oup.com/genetics/article/215/3/779/5930459">Ralph et al.&nbsp;(2020)</a> for more detail.</p>
</div></div></section>
<section id="what-if-we-need-mutations-though" class="level2">
<h2 class="anchored" data-anchor-id="what-if-we-need-mutations-though">What if we need mutations though?</h2>
<div class="fragment">
<p>Coalescent and mutation processes can be decoupled!</p>
<center>
<p><img src="images/tree_sequence_diagram_no_muts.png" class="img-fluid"></p>
</center>
</div>
</section>
<section id="what-if-we-need-mutations-though-1" class="level2">
<h2 class="anchored" data-anchor-id="what-if-we-need-mutations-though-1">What if we need mutations though?</h2>
<p>Coalescent and mutation processes can be decoupled!</p>
<center>
<p><img src="images/tree_sequence_diagram.png" class="img-fluid"></p>
<div class="fragment">
<h3 class="anchored">
With <em>slendr</em>, we can add mutations <em>after</em> the simulation using <code>ts_mutate()</code>.
</h3>
</div>
</center>
<!-- ## Tree-sequence simulation in practice -->
<!-- Let' say we have a compiled `model` object. -->
<!-- Then, we can simulate a tree sequence like this: -->
<!-- ```{r} -->
<!-- ts <- msprime(model, sequence_length = 10e6, recombination_rate = 1e-8) -->
<!-- ``` -->
<!-- But we can also simulate mutations like this: -->
<!-- ```{r} -->
<!-- ts <- -->
<!--   msprime(model, sequence_length = 10e6, recombination_rate = 1e-8) %>% -->
<!--   ts_mutate(mutation_rate = 1e-8) -->
<!-- ``` -->
<!-- <br> -->
<!-- <center>We will be using `ts_mutate()` throughout.</center> -->
<!-- ## Tree sequences are _very_ efficient -->
<!-- <br> -->
<!-- This simulates 2 $\times$ 10000 chromosomes of 100 Mb: -->
<!-- ```{r} -->
<!-- #| eval: false -->
<!-- pop <- population("pop", time = 1e6, N = 10000) -->
<!-- model <- compile_model(pop, generation_time = 30, direction = "backward") -->
<!-- ts <- msprime(model, sequence_length = 100e6, recombination_rate = 1e-8) -->
<!-- ``` -->
<!-- . . . -->
<!-- <br> -->
<!-- **Runs in less than 30 seconds on my laptop!** -->
<!-- **Takes about 66 Mb of memory!** -->
<!-- ## How does this work?! -->
<!-- . . . -->
<!-- <center> -->

<!-- ![](images/tables.jpeg) -->
<!-- <center> -->

<!-- ## Tree-sequence tables -->
<!-- ::: row -->
<!-- ::: columns -->
<!-- ::: {.column width="60%"} -->
<!-- A tree (sequence) can be represented by -->
<!-- ::: incremental -->

<!-- -   a table of <font color="orange">n</font><font color="green">o</font><font color="darkblue">d</font><font color="green">e</font><font color="darkblue">s</font>, -->
<!-- -   a table of [edges]{.underline} between nodes, -->
<!-- -   a table of <font color="red">mutations</font> on edges -->
<!-- ::: -->
<!-- ::: -->
<!-- ::: {.column width="40%"} -->
<!-- <br> -->
<!-- <center>![](images/tree_diagram.png)</center> -->
<!-- ::: -->
<!-- ::: -->
<!-- ::: -->
<!-- . . . -->
<!-- <center> -->

<!-- <h3>**A set of such tables is a tree sequence.**</h3> -->
<!-- </center> -->
<!-- ::: aside -->

<!-- You can find much more information in  ([tskit docs](https://tskit.dev/tutorials/tables_and_editing.html)) -->
<!-- ::: -->

<!-- ## Tree-sequence tables in practice -->
<!-- ::: columns -->
<!-- ::: {.column width="50%"} -->
<!-- ```{r} -->
<!-- #| echo: false -->
<!-- #| fig-height: 13 -->
<!-- set.seed(123) -->
<!-- ts <- msprime(model, sequence_length = 1e6, recombination_rate = 1e-8, random_seed = 42) %>% ts_mutate(1e-8, random_seed = 42) -->
<!-- # make a tiny example simplified tree sequence -->
<!-- ts_tiny <- ts_samples(ts) %>% sample_n(4) %>% pull(name) %>% ts_simplify(ts, simplify_to = .) -->
<!-- # extract tree #1 as an ape tree and also a tskit tree -->
<!-- t_phylo <- ts_phylo(ts_tiny, 1, quiet = TRUE) -->
<!-- t_tskit <- ts_tree(ts_tiny, 1) -->
<!-- # plot the phylo tree with ape -->
<!-- suppressPackageStartupMessages(library(ggtree)) -->
<!-- nodes <- ts_nodes(t_phylo) %>% as_tibble %>% dplyr::select(node = phylo_id, pop, node_id) -->
<!-- ggtree(t_phylo, branch.length = "none") %<+% nodes + -->
<!--   geom_label(aes(label = node_id), size = 15) +  -->
<!--   guides(color = "none") -->
<!-- ``` -->
<!-- ::: -->
<!-- ::: {.column width="50%"} -->
<!-- ::: fragment -->
<!-- nodes: -->
<!-- ```{r} -->
<!-- #| echo: false -->
<!-- ts_nodes(t_phylo) %>% head(3) %>% .[, c("node_id", "pop_id", "time")] %>% as.data.frame() -->
<!-- ``` -->
<!-- ::: -->
<!-- ::: fragment -->
<!-- edges: -->
<!-- ```{r} -->
<!-- #| echo: false -->
<!-- ts_edges(t_phylo) %>% head(3) %>% .[, c("child_node_id", "parent_node_id")] %>% as.data.frame() -->
<!-- ``` -->
<!-- ::: -->
<!-- ::: fragment -->
<!-- mutations: -->
<!-- ```{r} -->
<!-- #| echo: false -->

<!-- ts_table(ts_tiny, "mutations") %>% filter(node %in% c(53, 22, 20, 74, 9)) %>% head(3) %>% .[, c("id", "site", "node", "time")] %>% as.data.frame() -->
<!-- ``` -->
<!-- ::: -->
<!-- ::: -->
<!-- ::: -->
</section>
<section id="lets-take-the-model-from-earlier" class="level2">
<h2 class="anchored" data-anchor-id="lets-take-the-model-from-earlier">Let’s take the <code>model</code> from earlier…</h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2023_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="and-simulate-data-from-it" class="level2">
<h2 class="anchored" data-anchor-id="and-simulate-data-from-it">… and simulate data from it</h2>
<p><br></p>
<p>In our script we’ll have something like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1"></a><span class="sc">&lt;</span>... <span class="fu">population</span>() definitions ...<span class="sc">&gt;</span></span>
<span id="cb43-2"><a href="#cb43-2"></a></span>
<span id="cb43-3"><a href="#cb43-3"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(...)</span>
<span id="cb43-4"><a href="#cb43-4"></a>  </span>
<span id="cb43-5"><a href="#cb43-5"></a>ts <span class="ot">&lt;-</span></span>
<span id="cb43-6"><a href="#cb43-6"></a>  <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">50e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="and-simulate-data-from-it-1" class="level2">
<h2 class="anchored" data-anchor-id="and-simulate-data-from-it-1">… and simulate data from it</h2>
<p><br></p>
<p>In our script we’ll have something like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1"></a><span class="sc">&lt;</span>... <span class="fu">population</span>() definitions ...<span class="sc">&gt;</span></span>
<span id="cb44-2"><a href="#cb44-2"></a></span>
<span id="cb44-3"><a href="#cb44-3"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(...)</span>
<span id="cb44-4"><a href="#cb44-4"></a>  </span>
<span id="cb44-5"><a href="#cb44-5"></a>ts <span class="ot">&lt;-</span></span>
<span id="cb44-6"><a href="#cb44-6"></a>  <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">50e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span></span>
<span id="cb44-7"><a href="#cb44-7"></a>  <span class="fu">ts_mutate</span>(<span class="at">mutation_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment">
<p><br></p>
<center>
<h2 class="anchored">
Now we can simulate data.<br>How do we work with this <code>ts</code> thing?
</h2>
</center>
</div>
<!-- ## Conversion to other genotype formats -->
<!-- If you have a tree-sequence object `ts`, you can do... -->
<!-- ```{r} -->
<!-- #| eval: false -->
<!-- ts_vcf(ts, path = "path/to/a/file.vcf.gz") -->
<!-- ``` -->
<!-- ```{r} -->
<!-- #| eval: false -->
<!-- ts_eigenstrat(ts, prefix = "path/to/eigenstrat/prefix") -->
<!-- ``` -->
<!-- ```{r} -->
<!-- #| echo: false -->
<!-- ts <- ts_simplify(ts, simplify_to = c("pop1_1", "pop2_1")) -->
<!-- ``` -->
<!-- ```{r} -->
<!-- ts_genotypes(ts) -->
<!-- ``` -->
</section>
<section id="slendrs-r-interface-to-tskit-statistics" class="level2">
<h2 class="anchored" data-anchor-id="slendrs-r-interface-to-tskit-statistics"><em>slendr</em>’s R interface to <a href="https://tskit.dev/tskit"><em>tskit</em></a> statistics</h2>
<center>
<p><img src="images/slendr_tskit.png" class="img-fluid"></p>
</center>
<p>Allele-frequecy spectrum, nucleotide diversity <span class="math inline">\(\pi\)</span>, <span class="math inline">\(F_{ST}\)</span>, etc.</p>
</section>
<section id="extracting-sample-information" class="level2">
<h2 class="anchored" data-anchor-id="extracting-sample-information">Extracting sample information</h2>
<p>We can get samples recorded in <code>ts</code> with <code>ts_samples()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-output-location="fragment">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 3
  name    time pop  
  &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;
1 pop1_1     0 pop1 
2 pop1_2     0 pop1 
3 pop1_3     0 pop1 
4 pop1_4     0 pop1 
5 pop1_5     0 pop1 </code></pre>
</div>
</div>
<p>. . .</p>
<p>Sometimes a shortcut <code>ts_names()</code> can be useful:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_names</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pop1_1" "pop1_2" "pop1_3" "pop1_4" "pop1_5"</code></pre>
</div>
</div>
</section>
<section id="example-allele-frequency-spectrum" class="level2">
<h2 class="anchored" data-anchor-id="example-allele-frequency-spectrum">Example: <a href="https://en.wikipedia.org/wiki/Allele_frequency_spectrum">allele frequency spectrum</a></h2>
<div class="columns">
<div class="column" style="width:45%;">
<p>Sample 5 individuals:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>names <span class="ot">&lt;-</span> <span class="fu">ts_names</span>(ts)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>names</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pop_1" "pop_2" "pop_3" "pop_4" "pop_5"</code></pre>
</div>
</div>
<div class="fragment">
<p><br></p>
<p>Compute the AFS:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>afs <span class="ot">&lt;-</span> <span class="fu">ts_afs</span>(ts, <span class="fu">list</span>(names))</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>afs[<span class="sc">-</span><span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 3774 1887 1385 1054  701  666  553  438  469  423</code></pre>
</div>
</div>
</div>
</div><div class="column" style="width:2%;">
<p>&nbsp;</p>
</div><div class="column" style="width:53%;">
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(afs[<span class="sc">-</span><span class="dv">1</span>], <span class="at">type =</span> <span class="st">"b"</span>,</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"allele count bin"</span>,</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">"frequency"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output-display">
<p><img src="popgen2023_files/figure-html/unnamed-chunk-45-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</div>
</div>
</div>
<div class="fragment">
<p><small> <strong>Note:</strong> We drop the first element (<code>afs[-1]</code>) technical reasons related to <em>tskit</em>. You don’t have to worry about that here, but you can read <a href="https://tskit.dev/tutorials/analysing_tree_sequences.html#sec-tutorial-afs-zeroth-entry">this</a> for more detail. </small></p>
</div>
</section>
</section>
<section id="exercise-2" class="level1 page-columns page-full">
<h1>Exercise #2</h1>
<p><br><br><br><br></p>

<div class="no-row-height column-margin column-container"><div class="">
<p>Use <a href="https://bodkan.quarto.pub/ku-popgen2023-onepage/">these</a> one-page handouts for reference.</p>
</div></div><section id="exercise-2-estimating-n_e-from-afs" class="level2">
<h2 class="anchored" data-anchor-id="exercise-2-estimating-n_e-from-afs">Exercise #2: estimating <span class="math inline">\(N_e\)</span> from AFS</h2>
<p>Imagine you sequenced 10 samples from a population and computed this AFS vector (# singletons, doubletons, etc.):</p>
<!-- ```{r} -->
<!-- #| echo: false -->
<!-- dput(as.vector(observed_afs)) -->
<!-- ``` -->
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>afs_observed <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">2520</span>, <span class="dv">1449</span>, <span class="dv">855</span>, <span class="dv">622</span>, <span class="dv">530</span>, <span class="dv">446</span>, <span class="dv">365</span>, <span class="dv">334</span>, <span class="dv">349</span>, <span class="dv">244</span>,</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>                  <span class="dv">264</span>, <span class="dv">218</span>,  <span class="dv">133</span>, <span class="dv">173</span>, <span class="dv">159</span>, <span class="dv">142</span>, <span class="dv">167</span>, <span class="dv">129</span>, <span class="dv">125</span>, <span class="dv">143</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>. . .</p>
<p>You know that the population had constant <span class="math inline">\(N_e\)</span> somewhere between 1000 and 30000 for the past 100,000 generations, and had mutation and recombination rates of 1e-8.</p>
<p>. . .</p>
<hr>
<p><strong>Try to guess the true </strong> <span class="math inline">\(N_e\)</span> given the observed AFS by running single-population simulations for a range of <span class="math inline">\(N_e\)</span> values and comparing <code>ts_afs()</code> results to <code>afs_observed</code>.</p>
</section>
<section id="exercise-2-hints" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="exercise-2-hints">Exercise #2: hints</h2>
<ol type="1">
<li>Write a function (in a new script <code>model2.R</code>) taking <code>Ne</code> as a parameter. It should create one population, compile a model, run <code>ts_afs()</code> for 10 samples, and return the AFS.</li>
</ol>
<div class="fragment">
<p>If we call it <code>simulate_afs()</code>, it should work like this (your numbers will be different, of course!):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">simulate_afs</span>(<span class="at">Ne =</span> <span class="dv">1000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 4000 2000 1333 1000  800  667  571  500  444  400  364</code></pre>
</div>
</div>
</div>
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">simulate_afs</span>(<span class="at">Ne =</span> <span class="dv">12000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 48000 24000 16000 12000  9600  8000  6857  6000  5333  4800  4364</code></pre>
</div>
</div>
</div>

<div class="no-row-height column-margin column-container"><div class="">
<p>[Don’t panic—you can pretty much just tweak bits of my earlier example code.]</p>
</div></div></section>
<section id="exercise-2-hints-1" class="level2">
<h2 class="anchored" data-anchor-id="exercise-2-hints-1">Exercise #2: hints</h2>
<ol start="2" type="1">
<li><p>Using your custom <code>simulate_afs()</code> function, find the value of <code>Ne</code> that will give the closest AFS to the observed AFS:</p>
<ul>
<li><p><em>option #1</em> [<em>easy</em>]: Plot AFS vectors for various <span class="math inline">\(N_e\)</span> values, then eyeball which looks closest to the observed AFS</p></li>
<li><p><em>option #2</em> [<em>harder</em>]: Simulate AFS vectors in steps of possible <code>Ne</code> (maybe <code>lapply()</code>?), find the <a href="https://en.wikipedia.org/wiki/Mean_squared_error">closest</a> AFS</p></li>
</ul></li>
</ol>
</section>
<section id="exercise-2-hints-2" class="level2">
<h2 class="anchored" data-anchor-id="exercise-2-hints-2">Exercise #2: hints</h2>
<p>If you’re completely lost, you can start from this “template” again:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr)</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a><span class="fu">init_env</span>()</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>simulate_afs <span class="ot">&lt;-</span> <span class="cf">function</span>(Ne) {</span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a>  <span class="sc">&lt;</span>... AFS simulation code you cannibalized from my slides ...<span class="sc">&gt;</span></span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb59-7"><a href="#cb59-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-8"><a href="#cb59-8" aria-hidden="true" tabindex="-1"></a>afs_1k  <span class="ot">&lt;-</span> <span class="fu">simulate_afs</span>(<span class="at">Ne =</span> <span class="dv">1000</span>)</span>
<span id="cb59-9"><a href="#cb59-9" aria-hidden="true" tabindex="-1"></a>afs_12k <span class="ot">&lt;-</span> <span class="fu">simulate_afs</span>(<span class="at">Ne =</span> <span class="dv">12000</span>)</span>
<span id="cb59-10"><a href="#cb59-10" aria-hidden="true" tabindex="-1"></a><span class="sc">&lt;</span>...<span class="sc">&gt;</span></span>
<span id="cb59-11"><a href="#cb59-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-12"><a href="#cb59-12" aria-hidden="true" tabindex="-1"></a><span class="er">&lt;</span>... some plotting or statistics ...<span class="sc">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="exercise-2-solution" class="level1 page-columns page-full">
<h1>Exercise #2: solution</h1>
<p><br><br><br><br></p>

<div class="no-row-height column-margin column-container"><div class="">
<p>See <a href="https://github.com/bodkan/ku-popgen2023/blob/main/solutions/ex2_simple.R"><code>ex2_simple.R</code></a> for a simple “eyeballing” solution.<br></p>
<p>See <a href="https://github.com/bodkan/ku-popgen2023/blob/main/solutions/ex2_grid.R"><code>ex2_grid.R</code></a> for a more elaborate grid-search solution.</p>
</div></div></section>
<section id="gene-flow-admixture" class="level1">
<h1>Gene flow / admixture</h1>
<section id="we-use-the-function-gene_flow" class="level2">
<h2 class="anchored" data-anchor-id="we-use-the-function-gene_flow">We use the function <code>gene_flow()</code></h2>
<p><br></p>
<p>If we have populations <code>p1</code> and <code>p2</code>, we schedule gene flow as:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">&lt;-</span> <span class="fu">gene_flow</span>(<span class="at">from =</span> p1, <span class="at">to =</span> p2, <span class="at">start =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">600</span>, <span class="at">rate =</span> <span class="fl">0.13</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>. . .</p>
<p><br></p>
<p>Multiple gene-flow events can be gathered in a list:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gene_flow</span>(<span class="at">from =</span> p1, <span class="at">to =</span> p2, <span class="at">start =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">600</span>, <span class="at">rate =</span> <span class="fl">0.13</span>),</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gene_flow</span>(<span class="at">from =</span> ..., <span class="at">to =</span> ..., <span class="at">start =</span> ..., <span class="at">end =</span> ..., <span class="at">rate =</span> ...),</span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>  ...</span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>. . .</p>
<p><br></p>
<center>
<strong><code>gene_flow()</code> checks for consistency!</strong>
</center>
</section>
<section id="plugging-gene-flows-into-a-model" class="level2">
<h2 class="anchored" data-anchor-id="plugging-gene-flows-into-a-model">Plugging gene flow(s) into a model</h2>
<p><br> If you have a <em>slendr</em> script, you can do this: <br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1"></a><span class="sc">&lt;</span>... <span class="fu">population</span>() definitions ...<span class="sc">&gt;</span></span>
<span id="cb62-2"><a href="#cb62-2"></a></span>
<span id="cb62-3"><a href="#cb62-3"></a>gf <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb62-4"><a href="#cb62-4"></a>  <span class="fu">gene_flow</span>(<span class="at">from =</span> ... <span class="at">to =</span> ..., <span class="at">start =</span> ..., <span class="at">end =</span> ..., <span class="at">rate =</span> ...),</span>
<span id="cb62-5"><a href="#cb62-5"></a>  <span class="fu">gene_flow</span>(<span class="at">from =</span> ..., <span class="at">to =</span> ..., <span class="at">start =</span> ..., <span class="at">end =</span> ..., <span class="at">rate =</span> ...),</span>
<span id="cb62-6"><a href="#cb62-6"></a>  ...</span>
<span id="cb62-7"><a href="#cb62-7"></a>)</span>
<span id="cb62-8"><a href="#cb62-8"></a></span>
<span id="cb62-9"><a href="#cb62-9"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb62-10"><a href="#cb62-10"></a>  <span class="at">populations =</span> <span class="fu">list</span>(...),</span>
<span id="cb62-11"><a href="#cb62-11"></a>  <span class="at">gene_flow =</span> gf,           <span class="co"># &lt;--- gene flow(s) specified</span></span>
<span id="cb62-12"><a href="#cb62-12"></a>  <span class="at">generation_time =</span> <span class="sc">&lt;</span>...<span class="sc">&gt;</span></span>
<span id="cb62-13"><a href="#cb62-13"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="exercise-3" class="level1 page-columns page-full">
<h1>Exercise #3</h1>
<p><br><br><br><br></p>

<div class="no-row-height column-margin column-container"><div class="">
<p>Use <a href="https://bodkan.quarto.pub/ku-popgen2023-onepage/">these</a> one-page handouts for reference.</p>
</div></div><section id="exercise-3-adding-gene-flow" class="level2">
<h2 class="anchored" data-anchor-id="exercise-3-adding-gene-flow">Exercise #3: adding gene flow</h2>
<p>Add NEA → EUR introgression to your <code>model1.R</code> script:</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2023_files/figure-html/unnamed-chunk-55-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>&nbsp;</p>
</section>
<section id="exercise-3-adding-gene-flow-1" class="level2">
<h2 class="anchored" data-anchor-id="exercise-3-adding-gene-flow-1">Exercise #3: adding gene flow</h2>
<p>Add NEA → EUR introgression to your <code>model1.R</code> script:</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2023_files/figure-html/unnamed-chunk-56-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>3% pulse (<code>rate = 0.03</code>) between 55-50 kya (<code>start</code>-<code>end</code>)</p>
</section>
</section>
<section id="exercise-3-solution" class="level1 page-columns page-full">
<h1>Exercise #3: solution</h1>
<p><br><br><br><br></p>

<div class="no-row-height column-margin column-container"><div class="">
<p>See <a href="https://github.com/bodkan/ku-popgen2023/blob/main/solutions/ex3_ex4.R"><code>ex3_ex4.R</code> script on GitHub</a> for a solution.</p>
</div></div></section>
<section id="exercise-4" class="level1 page-columns page-full">
<h1>Exercise #4</h1>
<p><br><br><br><br></p>

<div class="no-row-height column-margin column-container"><div class="">
<p>Use <a href="https://bodkan.quarto.pub/ku-popgen2023-onepage/">these</a> one-page handouts for reference.</p>
</div></div><section id="exercise-4-more-statistics" class="level2">
<h2 class="anchored" data-anchor-id="exercise-4-more-statistics">Exercise #4: more statistics!</h2>
<p>Simulate a <code>ts</code> tree sequence from your complete Neanderthal introgression model in <code>model1.R</code>.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2023_files/figure-html/unnamed-chunk-57-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>(Remember to add mutations with <code>ts_mutate()</code>.)</p>
</section>
<section id="exercise-4-more-statistics-1" class="level2">
<h2 class="anchored" data-anchor-id="exercise-4-more-statistics-1">Exercise #4: more statistics!</h2>
<p><strong>Compute (some of) these on your tree sequence:</strong></p>
<ul>
<li><p>nucleotide <a href="https://www.slendr.net/reference/ts_diversity.html#ref-examples"><code>ts_diversity()</code></a> in each population <br></p></li>
<li><p><a href="https://www.slendr.net/reference/ts_divergence.html#ref-examples"><code>ts_divergence()</code></a> between populations<br></p></li>
<li><p><a href="https://www.slendr.net/reference/ts_fst.html#ref-examples"><code>ts_fst()</code></a> between populations<br></p></li>
<li><p>outgroup <a href="https://www.slendr.net/reference/ts_f4ratio.html#ref-examples"><code>ts_f3()</code></a> using CHIMP as the outgroup (A!) for different pairs of “B” and “C” populations<br></p></li>
<li><p><a href="https://www.slendr.net/reference/ts_f4ratio.html#ref-examples"><code>ts_f4()</code></a> test of Neanderthal introgression in EUR</p>
<p>[hint: <span class="math inline">\(f_4\)</span>(&lt;AFR&gt;, &lt;EUR&gt;; &lt;NEAND&gt;, &lt;CHIMP&gt;)]</p></li>
</ul>
<center>
<strong>Do your results fit expectations from your model?</strong>
</center>
</section>
<section id="exercise-4-hint" class="level2">
<h2 class="anchored" data-anchor-id="exercise-4-hint">Exercise #4: hint</h2>
<p>For multipopulation statistics, you need a (named) list of samples recorded in each population.</p>
<p><code>ts_names()</code> has a <code>split = "pop"</code> option just for this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">ts_names</span>(ts, <span class="at">split =</span> <span class="st">"pop"</span>)</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>samples</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>$AFR
[1] "AFR_1" "AFR_2" "AFR_3"

$NEA
[1] "NEA_1" "NEA_2" "NEA_3"</code></pre>
</div>
</div>
<p>You can use this in place of <code>sample_names</code> in code like:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_diversity</span>(ts, <span class="at">sample_sets =</span> samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="exercise-4-solution" class="level1 page-columns page-full">
<h1>Exercise #4: solution</h1>
<p><br><br><br><br></p>

<div class="no-row-height column-margin column-container"><div class="">
<p>See <a href="https://github.com/bodkan/ku-popgen2023/blob/main/solutions/ex3_ex4.R"><code>ex3_ex4.R</code> script on GitHub</a> for a solution.</p>
</div></div><section id="more-information" class="level2">
<h2 class="anchored" data-anchor-id="more-information">More information</h2>
<p><br></p>
<ul>
<li><p><em>slendr</em> paper is now in <a href="https://evolbiol.peercommunityin.org/articles/rec?id=630">PCI EvolBiol</a></p></li>
<li><p>documentation, tutorials is <a href="https://www.slendr.net">here</a></p></li>
<li><p>GitHub repo (bug reports!) is <a href="http://github.com/bodkan/slendr">here</a></p></li>
</ul>
<hr>
<ul>
<li>check out my new <a href="http://github.com/bodkan/demografr"><em>demografr</em></a> inference package</li>
</ul>
<hr>
<ul>
<li>contact details at <a href="https://bodkan.net">bodkan.net</a></li>
</ul>
</section>
</section>
<section id="section-9" class="level1">
<h1></h1>
<h1>
<p><font color="orange">B</font><font color="darkblue">O</font><font color="green">N</font><font color="red">U</font><font color="pink">S</font> <font color="purple">H</font><font color="orange">O</font><font color="green">M</font><font color="lightblue">E</font><font color="cyan">W</font><font color="green">O</font><font color="red">R</font><font color="orange">K</font></p>
</h1></section>
<section id="time-series-data" class="level1">
<h1>Time-series data</h1>
<section id="sampling-adna-samples-through-time" class="level2">
<h2 class="anchored" data-anchor-id="sampling-adna-samples-through-time">Sampling aDNA samples through time</h2>
<p><br></p>
<p>Imagine we have <code>pop1</code>, <code>pop2</code>, … compiled in a <code>model</code>.</p>
<p><br></p>
<p>To record <em>ancient</em> individuals in the tree sequence, we can use <code>schedule_sampling()</code> like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="fu">schedule_sampling</span>(</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>  model,                <span class="co"># compiled slendr model object</span></span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">times =</span> <span class="fu">c</span>(<span class="dv">100</span>, <span class="dv">500</span>),  <span class="co"># at these times (can be also a single number) ...</span></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(pop1, <span class="dv">42</span>),       <span class="co"># ... sample 42 individuals from pop1</span></span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(pop2, <span class="dv">10</span>),       <span class="co"># ... sample 10 individuals from pop2</span></span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(pop3, <span class="dv">1</span>)         <span class="co"># ... sample 1 individual from pop 3</span></span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="sampling-schedule-format" class="level2">
<h2 class="anchored" data-anchor-id="sampling-schedule-format">Sampling schedule format</h2>
<p>The output of <code>schedule_sampling()</code> is a plain data frame:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="fu">c</span>(<span class="dv">40000</span>, <span class="dv">30000</span>, <span class="dv">20000</span>, <span class="dv">10000</span>), <span class="fu">list</span>(eur, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 7
   time pop       n y_orig x_orig y     x    
  &lt;int&gt; &lt;chr&gt; &lt;int&gt; &lt;lgl&gt;  &lt;lgl&gt;  &lt;lgl&gt; &lt;lgl&gt;
1 10000 EUR       1 NA     NA     NA    NA   
2 20000 EUR       1 NA     NA     NA    NA   
3 30000 EUR       1 NA     NA     NA    NA   
4 40000 EUR       1 NA     NA     NA    NA   </code></pre>
</div>
</div>
<p>. . .</p>
<p>We can bind multiple sampling schedules together, giving us finer control about sampling:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>eur_samples <span class="ot">&lt;-</span> <span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="fu">c</span>(<span class="dv">40000</span>, <span class="dv">30000</span>, <span class="dv">20000</span>, <span class="dv">10000</span>, <span class="dv">0</span>), <span class="fu">list</span>(eur, <span class="dv">1</span>))</span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>afr_samples <span class="ot">&lt;-</span> <span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="dv">0</span>, <span class="fu">list</span>(afr, <span class="dv">1</span>))</span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(eur_samples, afr_samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="how-to-use-a-sampling-schedule" class="level2">
<h2 class="anchored" data-anchor-id="how-to-use-a-sampling-schedule">How to use a sampling schedule?</h2>
<p>To sample individuals based on a given schedule, we use the <code>samples =</code> argument of the <code>msprime()</code> functions:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">samples =</span> samples, <span class="at">sequence_length =</span> <span class="fl">1e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_mutate</span>(<span class="at">mutation_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>. . .</p>
<p>We can verify that only specific individuals are recorded:</p>
<div class="columns">
<div class="column" style="width:40%;">
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  name   time pop  
  &lt;chr&gt; &lt;dbl&gt; &lt;chr&gt;
1 EUR_1 40000 EUR  
2 EUR_2 30000 EUR  
3 EUR_3 20000 EUR  
4 EUR_4 10000 EUR  
5 AFR_1     0 AFR  
6 EUR_5     0 EUR  </code></pre>
</div>
</div>
</div><div class="column" style="width:60%;">
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>ts</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>╔═══════════════════════════╗
║TreeSequence               ║
╠═══════════════╤═══════════╣
║Trees          │       1279║
╟───────────────┼───────────╢
║Sequence Length│    1000000║
╟───────────────┼───────────╢
║Time Units     │generations║
╟───────────────┼───────────╢
║Sample Nodes   │         12║
╟───────────────┼───────────╢
║Total Size     │  280.9 KiB║
╚═══════════════╧═══════════╝
╔═══════════╤════╤═════════╤════════════╗
║Table      │Rows│Size     │Has Metadata║
╠═══════════╪════╪═════════╪════════════╣
║Edges      │3927│122.7 KiB│          No║
╟───────────┼────┼─────────┼────────────╢
║Individuals│   6│192 Bytes│          No║
╟───────────┼────┼─────────┼────────────╢
║Migrations │   0│  8 Bytes│          No║
╟───────────┼────┼─────────┼────────────╢
║Mutations  │1617│ 58.4 KiB│          No║
╟───────────┼────┼─────────┼────────────╢
║Nodes      │ 913│ 25.0 KiB│          No║
╟───────────┼────┼─────────┼────────────╢
║Populations│   4│341 Bytes│         Yes║
╟───────────┼────┼─────────┼────────────╢
║Provenances│   2│  3.5 KiB│          No║
╟───────────┼────┼─────────┼────────────╢
║Sites      │1616│ 39.5 KiB│          No║
╚═══════════╧════╧═════════╧════════════╝</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="exercise-5" class="level1 page-columns page-full">
<h1>Exercise #5</h1>
<p><br><br><br><br></p>

<div class="no-row-height column-margin column-container"><div class="">
<p>Use <a href="https://bodkan.quarto.pub/ku-popgen2023-onepage/">these</a> one-page handouts for reference.</p>
</div></div><section id="exercise-5a-ancient-samples" class="level2">
<h2 class="anchored" data-anchor-id="exercise-5a-ancient-samples">Exercise #5a: ancient samples</h2>
<p>Let’s return to your introgression model:</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2023_files/figure-html/unnamed-chunk-67-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="exercise-5a-ancient-samples-1" class="level2">
<h2 class="anchored" data-anchor-id="exercise-5a-ancient-samples-1">Exercise #5a: ancient samples</h2>
<p>Simulate data from your model using this sampling:</p>
<ul>
<li>one present-day CHIMP and AFR individual</li>
<li>20 present-day EUR individuals</li>
<li>1 NEA at 70 ky, 1 NEA at 40 ky</li>
<li>1 EUR every 1000 years between 50-5 kya</li>
</ul>
<p>Reminder: you can do this by:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="co"># rbind(...) together individual schedule_sampling() data frames</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span></span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">msprime</span>(model, <span class="at">samples =</span> samples, <span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span></span>
<span id="cb75-5"><a href="#cb75-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_mutate</span>(<span class="at">mutation_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="exercise-5b-f_4-ratio-statistic" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="exercise-5b-f_4-ratio-statistic">Exercise #5b: <span class="math inline">\(f_4\)</span>-ratio statistic</h2>
<p>Use <span class="math inline">\(f_4\)</span>-ratio statistic to replicate the <a href="https://www.pnas.org/doi/10.1073/pnas.1814338116#fig01">following figure</a>:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/neand_decline.png" class="img-fluid figure-img" style="width:60.0%"></p>
</figure>
</div>

<div class="no-row-height column-margin column-container"><div class="">
<p>Hint: You can compute Neanderthal ancestry for a vector of individuals <code>X</code> as <code>ts_f4ratio(ts, X = X, "NEA_1", "NEA_2", "AFR_1", "CHIMP_1")</code>.</p>
</div></div></section>
</section>
<section id="exercise-5-solution" class="level1 page-columns page-full">
<h1>Exercise #5: solution</h1>
<p><br><br><br><br></p>

<div class="no-row-height column-margin column-container"><div class="">
<p>See <a href="https://github.com/bodkan/ku-popgen2023/blob/main/solutions/ex5.R"><code>ex5.R</code> script on GitHub</a> for a solution.</p>
</div></div><section id="standard-genotype-formats" class="level2">
<h2 class="anchored" data-anchor-id="standard-genotype-formats">Standard genotype formats</h2>
<p>If a tree sequence doesn’t cut it, you can always:</p>
<ul>
<li>export genotypes to a VCF file:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_vcf</span>(ts, <span class="at">path =</span> <span class="st">"path/to/a/file.vcf.gz"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>export genotypes in an EIGENSTRAT format:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_eigenstrat</span>(ts, <span class="at">prefix =</span> <span class="st">"path/to/eigenstrat/prefix"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>access genotypes in a data frame:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_genotypes</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-output-location="fragment">
<div class="cell-output cell-output-stdout">
<pre><code>   pos EUR_1_chr1 EUR_1_chr2 EUR_2_chr1 EUR_2_chr2 AFR_1_chr1 AFR_1_chr2
1 2945          0          0          1          0          0          0
2 3307          0          0          0          1          0          0
3 3535          0          0          0          1          1          0
4 5137          0          0          1          0          0          0
5 6014          0          0          0          1          1          0</code></pre>
</div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>