<!DOCTYPE html>
<html lang="en"><head>
<script src="popgen2023_files/libs/clipboard/clipboard.min.js"></script>
<script src="popgen2023_files/libs/quarto-html/tabby.min.js"></script>
<script src="popgen2023_files/libs/quarto-html/popper.min.js"></script>
<script src="popgen2023_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="popgen2023_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="popgen2023_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="popgen2023_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="popgen2023_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.3.433">

  <meta name="author" content="Martin Petr">
  <meta name="author" content="CC BY 4.0">
  <title>Simulations in population genetics</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="popgen2023_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="popgen2023_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="popgen2023_files/libs/revealjs/dist/theme/quarto.css">
  <link href="popgen2023_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="popgen2023_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="popgen2023_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="popgen2023_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Simulations in population genetics</h1>
  <p class="subtitle"><a href="http://www.popgen.dk/popgen23/">PopGen summer course 2023</a></p>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Martin Petr 
</div>
</div>
<div class="quarto-title-author">
<div class="quarto-title-author-name">
<a href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a> 
</div>
</div>
</div>

  <p class="date">August 2023</p>
</section>
<section id="first-things-first" class="title-slide slide level1 center">
<h1>First things first</h1>
<p>These slides and other resources are (and always will be) at:</p>
<p><a href="https://github.com/bodkan/ku-popgen2023">github.com/bodkan/ku-popgen2023</a></p>
<p>[a one-page version of the slides, exercise solutions, etc.]</p>
<p><br></p>
<p><strong>Open this link now so you have everything at hand later.</strong></p>
</section>

<section id="first-things-first-1" class="title-slide slide level1 center">
<h1>First things first</h1>
<p>In order to make sure you can do the exercises, you should log into your <a href="emily.popgen.dk:3838/">online RStudio session</a> and run this in your R console:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">setup_env</span>(<span class="at">agree =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<p>If you want to do the exercises on your own laptop (if you <em>can</em>, you <em>should</em>), you need to do this first before running the above:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"slendr"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>

<section id="section" class="title-slide slide level1 center">
<h1></h1>
<div class="columns">
<div class="column" style="width:70%;">
<blockquote>
<p>Many problems in population genetics cannot be solved by a mathematician, no matter how gifted. [It] is already clear that computer methods are very powerful. This is good. It […] <strong>permits people with limited mathematical knowledge to work on important problems</strong> […].</p>
</blockquote>
</div><div class="column" style="width:30%;">
<p><img data-src="images/crow.jpeg"></p>
<p><a href="https://en.wikipedia.org/wiki/James_F._Crow">James F. Crow</a> – <a href="http://www.gnxp.com/blog/2006/06/10-questions-for-jim-crow.php">interview</a></p>
</div>
</div>
</section>

<section>
<section id="why-use-simulations" class="title-slide slide level1 center">
<h1>Why use simulations?</h1>
<ol type="1">
<li>Making sense of inferred statistics</li>
<li>Fitting model parameters (i.e.&nbsp;<a href="https://en.wikipedia.org/wiki/Approximate_Bayesian_computation">ABC</a>)</li>
<li>Ground truth for method development</li>
</ol>
</section>
<section id="making-sense-of-inferred-statistics" class="slide level2">
<h2>Making sense of inferred statistics</h2>
<center>
<img data-src="images/fstats_sims.png">
</center>

<aside><div>
<p>Image from <a href="https://academic.oup.com/genetics/article/202/4/1485/5930214">Peter (2016)</a></p>
</div></aside></section>
<section id="fitting-model-parameters-i.e.-abc" class="slide level2">
<h2>Fitting model parameters (i.e.&nbsp;<a href="https://en.wikipedia.org/wiki/Approximate_Bayesian_computation">ABC</a>)</h2>
<center>
<img data-src="images/abc_scheme.png" style="width:50.0%">
</center>

<aside><div>
<p>Image from <a href="https://en.wikipedia.org/wiki/Approximate_Bayesian_computation">Wikipedia on ABC</a></p>
</div></aside></section>
<section id="ground-truth-for-method-development" class="slide level2">
<h2>Ground truth for method development</h2>
<center>
<img data-src="images/mcmc.png">
</center>

<aside><div>
<p>Image from <a href="https://www.nature.com/articles/ng.3015">Schiffels and Durbin (2014)</a></p>
</div></aside></section></section>
<section id="what-does-it-mean-to-simulate-a-genome" class="title-slide slide level1 center">
<h1>What does it mean to simulate a genome?</h1>
<p><br> How would you design an algorithm for a popgen simulation?</p>

<p><br>What minimum components are needed?</p>
</section>

<section id="if-we-want-to-simulate-population-genetics" class="title-slide slide level1 center">
<h1>If we want to simulate population genetics</h1>
<p><br></p>
<div class="fragment">
<p>We need <em>populations</em>.</p>
</div>
<div class="fragment">
<p>We need <em>genetics</em>.</p>
</div>
</section>

<section id="a-chromosome-is" class="title-slide slide level1 center">
<h1>A chromosome is…</h1>
<p><br></p>
<div class="fragment">
<p>…a linear sequence of nucleotides</p>
<ul>
<li>a list of characters (A/G/C/T nucleotides)</li>
<li>a list of 0 or 1 values (ancestral/derived allele),</li>
</ul>
</div>
<div class="fragment">
<p>which mutates at a given <em>mutation rate</em>,<br></p>
<p>and recombines at a certain <em>recombination rate</em>.</p>
</div>
</section>

<section id="a-population-is" class="title-slide slide level1 center">
<h1>A population is…</h1>
<div class="fragment">
<p><br></p>
<p>… a collection of <em>individuals</em> at a given point in time,<br><br> each carrying <em>chromosomes</em> inherited from its parents.</p>
</div>
</section>

<section>
<section id="section-1" class="title-slide slide level1 center">
<h1></h1>
<h1>
Single-locus<br>Wright-Fisher<br>simulation in R
</h1>
<p><br></p>
<h3>
“What is the expected trajectory of a neutral allele under the influence of genetic drift?”
</h3>
</section>
<section id="complete-simulation-r-script" class="slide level2">
<h2>Complete simulation R script</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb3" data-code-line-numbers="|1-3|5-6|8-9|10-11|13-14|17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>p_0 <span class="ot">&lt;-</span> <span class="fl">0.5</span>  <span class="co"># initial allele frequency ("50% marbles" from lecture #1)</span></span>
<span id="cb3-2"><a href="#cb3-2"></a>N   <span class="ot">&lt;-</span> <span class="dv">500</span>  <span class="co"># number of chromosomes in a population</span></span>
<span id="cb3-3"><a href="#cb3-3"></a>T   <span class="ot">&lt;-</span> <span class="dv">500</span>  <span class="co"># number of generations to simulate</span></span>
<span id="cb3-4"><a href="#cb3-4"></a></span>
<span id="cb3-5"><a href="#cb3-5"></a><span class="co"># a vector for storing frequencies over time</span></span>
<span id="cb3-6"><a href="#cb3-6"></a>p_trajectory <span class="ot">&lt;-</span> p_0 </span>
<span id="cb3-7"><a href="#cb3-7"></a></span>
<span id="cb3-8"><a href="#cb3-8"></a><span class="co"># in each generation:</span></span>
<span id="cb3-9"><a href="#cb3-9"></a><span class="cf">for</span> (gen_i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>T) {</span>
<span id="cb3-10"><a href="#cb3-10"></a>  <span class="co"># get frequency in the previous generation</span></span>
<span id="cb3-11"><a href="#cb3-11"></a>  p_prev <span class="ot">&lt;-</span> p_trajectory[gen_i <span class="sc">-</span> <span class="dv">1</span>] </span>
<span id="cb3-12"><a href="#cb3-12"></a></span>
<span id="cb3-13"><a href="#cb3-13"></a>  <span class="co"># calculate new frequency ("sampling marbles from a jar")</span></span>
<span id="cb3-14"><a href="#cb3-14"></a>  p_trajectory[gen_i] <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="dv">1</span>, N, p_prev) <span class="sc">/</span> N </span>
<span id="cb3-15"><a href="#cb3-15"></a>}</span>
<span id="cb3-16"><a href="#cb3-16"></a></span>
<span id="cb3-17"><a href="#cb3-17"></a>p_trajectory</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="n-500-p_0-0.5" class="slide level2">
<h2><span class="math inline">\(N\)</span> = 500, <span class="math inline">\(p_0 = 0.5\)</span></h2>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(p_trajectory, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>),</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h =</span> p_0, <span class="at">lty =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">"red"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="popgen2023_files/figure-revealjs/unnamed-chunk-4-1.png" width="960" class="r-stretch"></section>
<section id="lets-make-it-a-function" class="slide level2">
<h2>Let’s make it a function</h2>
<p><strong>Input:</strong> <span class="math inline">\(p_0\)</span>, <span class="math inline">\(N\)</span>, and the number of generations</p>
<p><strong>Output:</strong> allele frequency trajectory vector</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5" data-code-line-numbers="|1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>simulate <span class="ot">&lt;-</span> <span class="cf">function</span>(p_0, N, T) {</span>
<span id="cb5-2"><a href="#cb5-2"></a>  p_trajectory <span class="ot">&lt;-</span> p_0</span>
<span id="cb5-3"><a href="#cb5-3"></a></span>
<span id="cb5-4"><a href="#cb5-4"></a>  <span class="cf">for</span> (gen_i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>T) {</span>
<span id="cb5-5"><a href="#cb5-5"></a>    p_prev <span class="ot">&lt;-</span> p_trajectory[gen_i <span class="sc">-</span> <span class="dv">1</span>]</span>
<span id="cb5-6"><a href="#cb5-6"></a>    p <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="dv">1</span>, N, p_prev) <span class="sc">/</span> N</span>
<span id="cb5-7"><a href="#cb5-7"></a>    p_trajectory[gen_i] <span class="ot">&lt;-</span> p</span>
<span id="cb5-8"><a href="#cb5-8"></a>  }</span>
<span id="cb5-9"><a href="#cb5-9"></a></span>
<span id="cb5-10"><a href="#cb5-10"></a>  p_trajectory</span>
<span id="cb5-11"><a href="#cb5-11"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="n-500-p_0-0.5-20-replicates" class="slide level2">
<h2><span class="math inline">\(N\)</span> = 500, <span class="math inline">\(p_0 = 0.5\)</span> (20 replicates)</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>reps <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">20</span>, <span class="fu">simulate</span>(<span class="at">N =</span> <span class="dv">500</span>, <span class="at">p_0 =</span> <span class="fl">0.5</span>, <span class="at">T =</span> <span class="dv">2000</span>))</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">matplot</span>(reps, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">lty =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="popgen2023_files/figure-revealjs/unnamed-chunk-6-1.png" width="960" class="r-stretch quarto-figure-center"></section>
<section id="n-1000-p_0-0.5-20-replicates" class="slide level2">
<h2><span class="math inline">\(N\)</span> = 1000, <span class="math inline">\(p_0 = 0.5\)</span> (20 replicates)</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>reps <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">20</span>, <span class="fu">simulate</span>(<span class="at">N =</span> <span class="dv">1000</span>, <span class="at">p_0 =</span> <span class="fl">0.5</span>, <span class="at">T =</span> <span class="dv">2000</span>))</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">matplot</span>(reps, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">lty =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="popgen2023_files/figure-revealjs/unnamed-chunk-7-1.png" width="960" class="r-stretch quarto-figure-center"></section>
<section id="n-5000-p_0-0.5-20-replicates" class="slide level2">
<h2><span class="math inline">\(N\)</span> = 5000, <span class="math inline">\(p_0 = 0.5\)</span> (20 replicates)</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>reps <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">20</span>, <span class="fu">simulate</span>(<span class="at">N =</span> <span class="dv">5000</span>, <span class="at">p_0 =</span> <span class="fl">0.5</span>, <span class="at">T =</span> <span class="dv">2000</span>))</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">matplot</span>(reps, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">lty =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="popgen2023_files/figure-revealjs/unnamed-chunk-8-1.png" width="960" class="r-stretch quarto-figure-center"></section>
<section id="n-10000-p_0-0.5-20-replicates" class="slide level2">
<h2><span class="math inline">\(N\)</span> = 10000, <span class="math inline">\(p_0 = 0.5\)</span> (20 replicates)</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>reps <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">20</span>, <span class="fu">simulate</span>(<span class="at">N =</span> <span class="dv">10000</span>, <span class="at">p_0 =</span> <span class="fl">0.5</span>, <span class="at">T =</span> <span class="dv">2000</span>))</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="fu">matplot</span>(reps, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">lty =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="popgen2023_files/figure-revealjs/unnamed-chunk-9-1.png" width="960" class="r-stretch quarto-figure-center"></section>
<section id="n-10000-p_0-0.5-100-replicates" class="slide level2">
<h2><span class="math inline">\(N\)</span> = 10000, <span class="math inline">\(p_0 = 0.5\)</span> (100 replicates)</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>reps <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">100</span>, <span class="fu">simulate</span>(<span class="at">N =</span> <span class="dv">10000</span>, <span class="at">p_0 =</span> <span class="fl">0.5</span>, <span class="at">T =</span> <span class="dv">30000</span>))</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="fu">matplot</span>(reps, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">lty =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="popgen2023_files/figure-revealjs/unnamed-chunk-10-1.png" width="960" class="r-stretch quarto-figure-center"></section>
<section id="n-10000-p_0-0.5-100-replicates-1" class="slide level2">
<h2><span class="math inline">\(N\)</span> = 10000, <span class="math inline">\(p_0 = 0.5\)</span> (100 replicates)</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>factors <span class="ot">&lt;-</span> MASS<span class="sc">::</span><span class="fu">fractions</span>(<span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">10</span>))</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">matplot</span>(reps, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">lty =</span> <span class="dv">1</span>)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> <span class="dv">10000</span> <span class="sc">*</span> factors, <span class="at">lwd =</span> <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="popgen2023_files/figure-revealjs/unnamed-chunk-11-1.png" width="960" class="r-stretch quarto-figure-center"></section>
<section id="expected-allele-frequency-distribution" class="slide level2">
<h2>Expected allele frequency distribution</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(parallel)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="st">"diffusion.rds"</span>)) {</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>p_0 <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>N <span class="ot">&lt;-</span> <span class="dv">10000</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>factors <span class="ot">&lt;-</span> MASS<span class="sc">::</span><span class="fu">fractions</span>(<span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">10</span>))</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>final_frequencies <span class="ot">&lt;-</span> parallel<span class="sc">::</span><span class="fu">mclapply</span>(</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">seq_along</span>(factors), <span class="cf">function</span>(i) {</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>    f <span class="ot">&lt;-</span> factors[i]</span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>    t <span class="ot">&lt;-</span> <span class="fu">as.integer</span>(N <span class="sc">*</span> f)</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>    <span class="co"># get complete trajectories as a matrix (each column a single trajectory)</span></span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>    reps <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">10000</span>, <span class="fu">simulate</span>(<span class="at">N =</span> N, <span class="at">p_0 =</span> p_0, <span class="at">T =</span> t))</span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>    <span class="co"># only keep the last slice of the matrix with the final frequencies</span></span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>(</span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>      <span class="at">t =</span> <span class="fu">sprintf</span>(<span class="st">"t = %s * N"</span>, f),</span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a>      <span class="at">freq =</span> reps[t, ]</span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a>  },</span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a>  <span class="at">mc.cores =</span> parallel<span class="sc">::</span><span class="fu">detectCores</span>()</span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span> <span class="fu">do.call</span>(rbind, .)</span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a>final_frequencies<span class="sc">$</span>t <span class="ot">&lt;-</span> forcats<span class="sc">::</span><span class="fu">fct_rev</span>(forcats<span class="sc">::</span><span class="fu">fct_relevel</span>(final_frequencies<span class="sc">$</span>t, <span class="fu">sprintf</span>(<span class="st">"t = %s * N"</span>, factors)))</span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(final_frequencies, <span class="st">"diffusion.rds"</span>)</span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a>  final_frequencies <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"diffusion.rds"</span>)</span>
<span id="cb12-28"><a href="#cb12-28" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb12-29"><a href="#cb12-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-30"><a href="#cb12-30" aria-hidden="true" tabindex="-1"></a>final_frequencies <span class="sc">%&gt;%</span> .[.<span class="sc">$</span>freq <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;</span> .<span class="sc">$</span>freq <span class="sc">&lt;</span> <span class="dv">1</span>, ] <span class="sc">%&gt;%</span></span>
<span id="cb12-31"><a href="#cb12-31" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb12-32"><a href="#cb12-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(freq, <span class="at">y =</span> <span class="fu">after_stat</span>(density), <span class="at">fill =</span> t), <span class="at">position =</span> <span class="st">"identity"</span>, <span class="at">bins =</span> <span class="dv">100</span>, <span class="at">alpha =</span> <span class="fl">0.75</span>) <span class="sc">+</span></span>
<span id="cb12-33"><a href="#cb12-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"allele frequency"</span>) <span class="sc">+</span></span>
<span id="cb12-34"><a href="#cb12-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">3</span>)) <span class="sc">+</span></span>
<span id="cb12-35"><a href="#cb12-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(t <span class="sc">~</span> .) <span class="sc">+</span></span>
<span id="cb12-36"><a href="#cb12-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">fill =</span> <span class="fu">guide_legend</span>(<span class="fu">sprintf</span>(<span class="st">"time since</span><span class="sc">\n</span><span class="st">the start</span><span class="sc">\n</span><span class="st">[assuming</span><span class="sc">\n</span><span class="st">N = %s]"</span>, N))) <span class="sc">+</span></span>
<span id="cb12-37"><a href="#cb12-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb12-38"><a href="#cb12-38" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">strip.text.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb12-39"><a href="#cb12-39" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>),</span>
<span id="cb12-40"><a href="#cb12-40" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>),</span>
<span id="cb12-41"><a href="#cb12-41" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>),</span>
<span id="cb12-42"><a href="#cb12-42" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>),</span>
<span id="cb12-43"><a href="#cb12-43" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>),</span>
<span id="cb12-44"><a href="#cb12-44" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="popgen2023_files/figure-revealjs/unnamed-chunk-12-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="fragment">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/kimura.jpeg"></p>
<figcaption><em>“Diffusion Models in Population Genetics”</em>, <a href="https://www.jstor.org/stable/3211856#metadata_info_tab_contents">Kimura (1964)</a></figcaption>
</figure>
</div>
</div>
</div>
</div>
</section></section>
<section id="section-2" class="title-slide slide level1 center" data-background-image="images/montypython.jpeg">
<h1></h1>
<div class="fragment">
<h1>
<br><br><br>
</h1>
<h1 color="black" style="background-color: white">
<p>But now for something completely different.</p>
</h1>
</div>
</section>

<section>
<section id="section-3" class="title-slide slide level1 center" data-background-image="images/montypython.jpeg">
<h1></h1>
<h1>
<br><br><br>
</h1>
<h1 color="black" style="background-color: white">
<p>Let’s talk about<br><em>real</em> simulations.</p>
</h1>
</section>
<section id="there-are-many-simulation-tools" class="slide level2">
<h2>There are many simulation tools</h2>
<p><br></p>
<p>The most famous and widely used are <a href="https://messerlab.org/slim/">SLiM</a> and <a href="https://tskit.dev/msprime/docs/stable/intro.html"><em>msprime</em></a>.</p>
<p><br></p>
<p>They are <u>very</u> powerful but both require:</p>
<div class="fragment">
<ul>
<li>quite a bit of programming knowledge,</li>
<li>a lot of code for non-trivial simulations (🐛🪲🐜).</li>
</ul>
</div>
<div class="fragment">
<p><br></p>
<center>
<h3>
<strong>The exercises will focus on the <a href="http://www.slendr.net"><em>slendr</em></a></strong><br>popgen simulation toolkit for R.
</h3>
</center>
</div>
</section>
<section id="section-4" class="slide level2">
<h2></h2>
<div class="columns">
<div class="column" style="width:60%;">
<h2>
What is SLiM?
</h2>
<div class="fragment">
<ul>
<li><p><strong>Forward-time simulator</strong></p></li>
<li><p>It’s fully programmable!</p></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Massive library of functions for:
<ul>
<li>Demographic events</li>
<li>Various mating systems</li>
<li>Selection, quantitative traits, …</li>
</ul></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>&gt; 700 pages long <a href="https://github.com/MesserLab/SLiM/releases/download/v3.7.1/SLiM_Manual.pdf">manual</a>!</li>
</ul>
</div>
</div><div class="column" style="width:40%;">
<center>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/sim_sketches.001.png" style="width:100.0%"></p>
<figcaption>Modified from <a href="http://alexeidrummond.org/bayesian_phylo_lectures/lecture10/">Alexei Drummond</a></figcaption>
</figure>
</div>
</center>
</div>
</div>
</section>
<section id="simple-neutral-simulation-in-slim" class="slide level2">
<h2>Simple neutral simulation in SLiM</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>initialize() {
    // create a neutral mutation type
    initializeMutationType("m1", 0.5, "f", 0.0);
    // initialize 1Mb segment
    initializeGenomicElementType("g1", m1, 1.0);
    initializeGenomicElement(g1, 0, 999999);
    // set mutation rate and recombination rate of the segment
    initializeMutationRate(1e-8);
    initializeRecombinationRate(1e-8);
}

// create an ancestral population p1 of 10000 diploid individuals
1 early() { sim.addSubpop("p1", 10000); }

// in generation 1000, create two daughter populations p2 and p3
1000 early() {
    sim.addSubpopSplit("p2", 5000, p1);
    sim.addSubpopSplit("p3", 1000, p1);
}

// in generation 10000, stop the simulation and save 100 individuals
// from p2 and p3 to a VCF file
10000 late() {
    p2_subset = sample(p2.individuals, 100);
    p3_subset = sample(p3.individuals, 100);
    c(p2_subset, p3_subset).genomes.outputVCF("/tmp/slim_output.vcf.gz");
    sim.simulationFinished();
    catn("DONE!");
}</code></pre>
</div>
</div>
</section>
<section id="section-5" class="slide level2">
<h2></h2>
<div class="columns">
<div class="column" style="width:60%;">
<h2>
What is <em>msprime</em>
</h2>
<div class="fragment">
<ul>
<li>A Python module for writing <strong>coalescent simulations</strong></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Extremely fast (genome-scale, population-scale data)</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>You must know Python fairly well to build complex models</li>
</ul>
</div>
</div><div class="column" style="width:40%;">
<center>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/sim_sketches.002.png" style="width:100.0%"></p>
<figcaption>Modified from <a href="http://alexeidrummond.org/bayesian_phylo_lectures/lecture10/">Alexei Drummond</a></figcaption>
</figure>
</div>
</center>
</div>
</div>
</section>
<section id="simple-simulation-using-msprime" class="slide level2">
<h2>Simple simulation using <em>msprime</em></h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>import msprime

demography = msprime.Demography()
demography.add_population(name="A", initial_size=10_000)
demography.add_population(name="B", initial_size=5_000)
demography.add_population(name="C", initial_size=1_000)
demography.add_population_split(time=1000, derived=["A", "B"], ancestral="C")

ts = msprime.sim_ancestry(
  sequence_length=10e6,
  recombination_rate=1e-8,
  samples={"A": 100, "B": 100},
  demography=demography
)</code></pre>
</div>
</div>
<p>source: <a href="https://tskit.dev/msprime/docs/stable/demography.html#demographic-models">link</a></p>
</section></section>
<section>
<section id="section-6" class="title-slide slide level1 center">
<h1></h1>
<center>
<h2>
<a href="https://www.slendr.net">www.slendr.net</a>
</h2>
<p><img data-src="images/slendr_logo.png" style="width:30.0%"></p>
</center>
<div class="fragment">
<center>
<h2>
Why a new package?
</h2>
</center>
</div>
</section>
<section id="spatial-simulations" class="slide level2">
<h2>Spatial simulations!</h2>
<center>
<img data-src="images/animation.gif" style="width:70.0%">
</center>
</section>
<section id="why-a-new-package" class="slide level2">
<h2>Why a new package?</h2>
<div class="fragment">
<ul>
<li><p>Most researchers are not expert programmers</p></li>
<li><p>All but the most trivial simulations require lots of code</p></li>
</ul>
</div>
<div class="fragment">
<ul>
<li><p>90% <sup><font color="blue">[citation needed]</font></sup> of simulations are basically the same!</p>
<ul>
<li><p>create populations (splits and <span class="math inline">\(N_e\)</span> changes)</p></li>
<li><p>specify if/how they should mix (rates and times)</p></li>
</ul></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Lot of code duplication across projects</li>
</ul>
</div>
<div class="fragment">
<center>
<strong><em>slendr</em></strong> <strong>makes “standard” simulations trivial and<br> unlocks new kinds of spatial simulations</strong>
</center>
</div>
</section></section>
<section>
<section id="lets-get-started" class="title-slide slide level1 center">
<h1>Let’s get started</h1>

</section>
<section id="everything-we-do-will-be-in-r" class="slide level2">
<h2>Everything we do will be in R</h2>
<p><br></p>
<p>Always start your R scripts with this (*):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">init_env</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<p>My solutions will also use these two packages:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<aside><div>
<p>(*) You can safely ignore the message about missing SLiM.</p>
</div></aside></section>
<section id="typical-steps-of-a-slendr-r-workflow" class="slide level2">
<h2>Typical steps of a <em>slendr</em> R workflow</h2>
<p><br></p>
<ol type="1">
<li>creating populations</li>
<li>scheduling population splits</li>
<li>programming <span class="math inline">\(N_e\)</span> size changes</li>
<li>encoding gene-flow events</li>
<li>simulation sequence of a given size</li>
<li>computing statistics from simulated outputs</li>
</ol>
</section>
<section id="creating-a-population" class="slide level2">
<h2>Creating a <code>population()</code></h2>
<p>Each needs a name, size and time of appearance (i.e., “split”):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment">
<p><br></p>
<p>This creates a normal R object. Typing it out gives a summary:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>pop1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'population' object 
-------------------------- 
name: pop1 
non-spatial population
stays until the end of the simulation

population history overview:
  - time 1: created as an ancestral population (N = 1000)</code></pre>
</div>
</div>
</div>
</section>
<section id="programming-population-splits" class="slide level2">
<h2>Programming population splits</h2>
<p>Splits are indicated by the <code>parent = &lt;pop&gt;</code> argument:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>pop2 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">50</span>, <span class="at">parent =</span> pop1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment">
<p><br></p>
<p>The split is reported in the “historical summary”:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>pop2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'population' object 
-------------------------- 
name: pop2 
non-spatial population
stays until the end of the simulation

population history overview:
  - time 50: split from pop1 (N = 100)</code></pre>
</div>
</div>
</div>
</section>
<section id="scheduling-resize-events-resize" class="slide level2">
<h2>Scheduling resize events – <code>resize()</code></h2>
<p>Step size decrease:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23" data-code-line-numbers="2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span>
<span id="cb23-2"><a href="#cb23-2"></a>pop1_step <span class="ot">&lt;-</span> <span class="fu">resize</span>(pop1, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">how =</span> <span class="st">"step"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<p>Exponential increase:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24" data-code-line-numbers="2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a>pop2 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">50</span>, <span class="at">parent =</span> pop1)</span>
<span id="cb24-2"><a href="#cb24-2"></a>pop2_exp <span class="ot">&lt;-</span> <span class="fu">resize</span>(pop2, <span class="at">N =</span> <span class="dv">10000</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">2000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="tidyverse-style-pipe-interface" class="slide level2">
<h2>Tidyverse-style <a href="https://magrittr.tidyverse.org">pipe</a> <code>%&gt;%</code> interface</h2>
<p>A more concise way to express the same thing as before.</p>
<p><br></p>
<p>Step size decrease:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>pop1 <span class="ot">&lt;-</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">how =</span> <span class="st">"step"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Exponential increase:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>pop2 <span class="ot">&lt;-</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">10000</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">2000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="a-more-complex-model" class="slide level2">
<h2>A more complex model</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>pop2 <span class="ot">&lt;-</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">300</span>, <span class="at">parent =</span> pop1) <span class="sc">%&gt;%</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">100</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">1000</span>)</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>pop3 <span class="ot">&lt;-</span></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop3"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">400</span>, <span class="at">parent =</span> pop2) <span class="sc">%&gt;%</span></span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">2500</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">800</span>)</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>pop4 <span class="ot">&lt;-</span></span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop4"</span>, <span class="at">N =</span> <span class="dv">1500</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">parent =</span> pop3) <span class="sc">%&gt;%</span></span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">700</span>, <span class="at">how =</span> <span class="st">"exponential"</span>, <span class="at">time =</span> <span class="dv">1200</span>, <span class="at">end =</span> <span class="dv">2000</span>)</span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a>pop5 <span class="ot">&lt;-</span></span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop5"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">600</span>, <span class="at">parent =</span> pop4) <span class="sc">%&gt;%</span></span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">50</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">900</span>) <span class="sc">%&gt;%</span></span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">1000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>, <span class="at">time =</span> <span class="dv">1600</span>, <span class="at">end =</span> <span class="dv">2200</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="each-object-carries-its-history" class="slide level2">
<h2>Each object carries its history!</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>pop5</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'population' object 
-------------------------- 
name: pop5 
non-spatial population
stays until the end of the simulation

population history overview:
  - time 600: split from pop4 (N = 100)
  - time 900: resize from 100 to 50 individuals
  - time 1600-2200: exponential resize from 50 to 1000 individuals</code></pre>
</div>
</div>
</section>
<section id="last-step-before-simulation-compile_model" class="slide level2">
<h2>Last step before simulation: <code>compile_model()</code></h2>
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(pop1, pop2, pop3, pop4, pop5),</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">generation_time =</span> <span class="dv">1</span>,</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">simulation_length =</span> <span class="dv">3000</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<center>
<strong>Compilation takes a list of model components, performs internal consistency checks, returns a single model object.</strong>
</center>
</section>
<section id="model-summary" class="slide level2">
<h2>Model summary</h2>
<p>Typing the compiled <code>model</code> into R prints a brief summary:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>model</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'model' object 
--------------------- 
populations: pop1, pop2, pop3, pop4, pop5 
geneflow events: [no geneflow]
generation time: 1 
time direction: forward 
total running length: 3000 model time units
model type: non-spatial

configuration files in: /private/var/folders/d_/hblb15pd3b94rg0v35920wd80000gn/T/RtmpebZ1Dn/fileb072f2d6afb </code></pre>
</div>
</div>
</section>
<section id="model-visualization" class="slide level2">
<h2>Model visualization</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="popgen2023_files/figure-revealjs/unnamed-chunk-29-1.png" width="672" class="r-stretch quarto-figure-center"></section></section>
<section>
<section id="exercise-1" class="title-slide slide level1 center">
<h1>Exercise #1</h1>
<p><br><br><br><br></p>

<aside><div>
<p>Use <a href="https://bodkan.quarto.pub/ku-popgen2023-onepage/">these</a> one-page handouts for reference.</p>
</div></aside></section>
<section id="workaround-for-an-rstudio-bug" class="slide level2">
<h2>Workaround for an RStudio bug</h2>
<div class="columns">
<div class="column" style="width:40%;">
<p>RStudio sometimes interferes with Python which is needed for simulations.</p>
<p>Go to <code>Tools</code> → <code>Global Options</code> in your RStudio and set the options like this.</p>
<p>Also happens on the <a href="http://emily.popgen.dk:3838/">shared server</a>!</p>
</div><div class="column" style="width:60%;">
<center>
<img data-src="images/rstudio_setting.png">
</center>
</div>
</div>
</section>
<section id="exercise-1-write-this-model-in-slendr" class="slide level2">
<h2>Exercise #1: write this model in <em>slendr</em></h2>
<div class="columns">
<div class="column" style="width:45%;">
<p><img data-src="images/intro_model1.png"></p>
</div><div class="column" style="width:55%;">
<div class="fragment">
<p>Start a <code>model1.R</code> script in <a href="http://emily.popgen.dk:3838/">RStudio</a> with this “template”:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="fu">init_env</span>()</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>chimp <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"CHIMP"</span>, <span class="at">time =</span> <span class="fl">7e6</span>, <span class="at">N =</span> <span class="dv">5000</span>)</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a><span class="co"># &lt;... rest of your code ...&gt;</span></span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">populations =</span> <span class="fu">list</span>(chimp, ...),</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">generation_time =</span> <span class="dv">30</span></span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model) <span class="co"># verify visually</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>

<aside><div>
<p>You can specify events in “units of years ago”. Just program times in a “decreasing order” (7Mya → 6Mya → …, as above).</p>
</div></aside></section></section>
<section id="exercise-1-solution" class="title-slide slide level1 center">
<h1>Exercise #1: solution</h1>
<p><br><br><br><br></p>

<aside><div>
<p>See <a href="https://github.com/bodkan/ku-popgen2023/blob/main/solutions/ex1.R"><code>ex1.R</code> script on GitHub</a> for a solution.</p>
</div></aside></section>

<section>
<section id="section-7" class="title-slide slide level1 center">
<h1></h1>
<h1>
So, now we can<br>write a <code>model</code>
</h1>
</section>
<section id="how-do-we-simulate-from-it" class="slide level2">
<h2>How do we simulate from it?</h2>
<p><em>slendr</em> has two built-in simulation “engine scripts”:</p>
<ul>
<li>SLiM engine (<a href="https://github.com/bodkan/slendr/blob/main/inst/scripts/script.slim">source</a>)</li>
<li><em>msprime</em> engine (<a href="https://github.com/bodkan/slendr/blob/main/inst/scripts/script.py">source</a>)</li>
</ul>
<p>They are designed to understand <em>slendr</em> models.</p>
<div class="fragment">
<p><br>All you need to simulate data is this one line of code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<center>
<p><br></p>
<h3>
<strong>You don’t have to write <em>msprime</em> or SLiM code!</strong>
</h3>
</center>
</div>
</section></section>
<section>
<section id="the-output-of-a-slendr-simulation-is-a-tree-sequence-ts" class="title-slide slide level1 center">
<h1>The output of a <em>slendr</em> simulation is a <strong>tree sequence (<code>ts</code>)</strong></h1>

</section>
<section id="what-is-tree-sequence" class="slide level2">
<h2>What is tree sequence?</h2>

<img data-src="images/tree_sequence_diagram_no_muts.png" style="width:80.0%" class="r-stretch quarto-figure-center"><ul>
<li>a record of full genetic ancestry of a set of samples</li>
<li>an encoding of DNA sequence carried by those samples</li>
<li>an efficient analysis framework</li>
</ul>
</section></section>
<section>
<section id="why-tree-sequence" class="title-slide slide level1 center">
<h1>Why tree sequence?</h1>
<p><br></p>
<h3>
Why not VCF or a normal genotype table?
</h3>
</section>
<section id="what-we-usually-have" class="slide level2">
<h2>What we usually have</h2>
<center>
<img data-src="images/vcf_screenshot.png" style="width:90.0%">
</center>
</section>
<section id="what-we-usually-want" class="slide level2">
<h2>What we usually <em>want</em></h2>
<p>An understanding of our samples’ evolutionary history:</p>
<center>
<img data-src="images/tree_sequence_diagram_no_muts.png">
</center>
<div class="fragment">
<center>
<h3>
<strong>This is exactly what a tree sequence <em>is</em>!</strong>
</h3>
</center>
</div>
</section>
<section id="the-magic-of-tree-sequences" class="slide level2">
<h2>The magic of tree sequences</h2>
<p>They allow computing of popgen statistics <em>without genotypes</em>!</p>
<center>
<p><img data-src="images/tree_sequence_diagram_no_muts.png"></p>
</center>
<p>There is a “duality” between mutations and branch lengths.</p>

<aside><div>
<p>See an amazing paper by <a href="https://academic.oup.com/genetics/article/215/3/779/5930459">Ralph et al.&nbsp;(2020)</a> for more detail.</p>
</div></aside></section>
<section id="what-if-we-need-mutations-though" class="slide level2">
<h2>What if we need mutations though?</h2>
<div class="fragment">
<p>Coalescent and mutation processes can be decoupled!</p>
<center>
<p><img data-src="images/tree_sequence_diagram_no_muts.png"></p>
</center>
</div>
</section>
<section id="what-if-we-need-mutations-though-1" class="slide level2">
<h2>What if we need mutations though?</h2>
<p>Coalescent and mutation processes can be decoupled!</p>
<center>
<p><img data-src="images/tree_sequence_diagram.png"></p>
<div class="fragment">
<h3>
With <em>slendr</em>, we can add mutations <em>after</em> the simulation using <code>ts_mutate()</code>.
</h3>
</div>
</center>
<!-- ## Tree-sequence simulation in practice -->
<!-- Let' say we have a compiled `model` object. -->
<!-- Then, we can simulate a tree sequence like this: -->
<!-- ```{r} -->
<!-- ts <- msprime(model, sequence_length = 10e6, recombination_rate = 1e-8) -->
<!-- ``` -->
<!-- But we can also simulate mutations like this: -->
<!-- ```{r} -->
<!-- ts <- -->
<!--   msprime(model, sequence_length = 10e6, recombination_rate = 1e-8) %>% -->
<!--   ts_mutate(mutation_rate = 1e-8) -->
<!-- ``` -->
<!-- <br> -->
<!-- <center>We will be using `ts_mutate()` throughout.</center> -->
<!-- ## Tree sequences are _very_ efficient -->
<!-- <br> -->
<!-- This simulates 2 $\times$ 10000 chromosomes of 100 Mb: -->
<!-- ```{r} -->
<!-- #| eval: false -->
<!-- pop <- population("pop", time = 1e6, N = 10000) -->
<!-- model <- compile_model(pop, generation_time = 30, direction = "backward") -->
<!-- ts <- msprime(model, sequence_length = 100e6, recombination_rate = 1e-8) -->
<!-- ``` -->
<!-- . . . -->
<!-- <br> -->
<!-- **Runs in less than 30 seconds on my laptop!** -->
<!-- **Takes about 66 Mb of memory!** -->
<!-- ## How does this work?! -->
<!-- . . . -->
<!-- <center> -->

<!-- ![](images/tables.jpeg) -->
<!-- <center> -->

<!-- ## Tree-sequence tables -->
<!-- ::: row -->
<!-- ::: columns -->
<!-- ::: {.column width="60%"} -->
<!-- A tree (sequence) can be represented by -->
<!-- ::: incremental -->

<!-- -   a table of <font color="orange">n</font><font color="green">o</font><font color="darkblue">d</font><font color="green">e</font><font color="darkblue">s</font>, -->
<!-- -   a table of [edges]{.underline} between nodes, -->
<!-- -   a table of <font color="red">mutations</font> on edges -->
<!-- ::: -->
<!-- ::: -->
<!-- ::: {.column width="40%"} -->
<!-- <br> -->
<!-- <center>![](images/tree_diagram.png)</center> -->
<!-- ::: -->
<!-- ::: -->
<!-- ::: -->
<!-- . . . -->
<!-- <center> -->

<!-- <h3>**A set of such tables is a tree sequence.**</h3> -->
<!-- </center> -->
<!-- ::: aside -->

<!-- You can find much more information in  ([tskit docs](https://tskit.dev/tutorials/tables_and_editing.html)) -->
<!-- ::: -->

<!-- ## Tree-sequence tables in practice -->
<!-- ::: columns -->
<!-- ::: {.column width="50%"} -->
<!-- ```{r} -->
<!-- #| echo: false -->
<!-- #| fig-height: 13 -->
<!-- set.seed(123) -->
<!-- ts <- msprime(model, sequence_length = 1e6, recombination_rate = 1e-8, random_seed = 42) %>% ts_mutate(1e-8, random_seed = 42) -->
<!-- # make a tiny example simplified tree sequence -->
<!-- ts_tiny <- ts_samples(ts) %>% sample_n(4) %>% pull(name) %>% ts_simplify(ts, simplify_to = .) -->
<!-- # extract tree #1 as an ape tree and also a tskit tree -->
<!-- t_phylo <- ts_phylo(ts_tiny, 1, quiet = TRUE) -->
<!-- t_tskit <- ts_tree(ts_tiny, 1) -->
<!-- # plot the phylo tree with ape -->
<!-- suppressPackageStartupMessages(library(ggtree)) -->
<!-- nodes <- ts_nodes(t_phylo) %>% as_tibble %>% dplyr::select(node = phylo_id, pop, node_id) -->
<!-- ggtree(t_phylo, branch.length = "none") %<+% nodes + -->
<!--   geom_label(aes(label = node_id), size = 15) +  -->
<!--   guides(color = "none") -->
<!-- ``` -->
<!-- ::: -->
<!-- ::: {.column width="50%"} -->
<!-- ::: fragment -->
<!-- nodes: -->
<!-- ```{r} -->
<!-- #| echo: false -->
<!-- ts_nodes(t_phylo) %>% head(3) %>% .[, c("node_id", "pop_id", "time")] %>% as.data.frame() -->
<!-- ``` -->
<!-- ::: -->
<!-- ::: fragment -->
<!-- edges: -->
<!-- ```{r} -->
<!-- #| echo: false -->
<!-- ts_edges(t_phylo) %>% head(3) %>% .[, c("child_node_id", "parent_node_id")] %>% as.data.frame() -->
<!-- ``` -->
<!-- ::: -->
<!-- ::: fragment -->
<!-- mutations: -->
<!-- ```{r} -->
<!-- #| echo: false -->

<!-- ts_table(ts_tiny, "mutations") %>% filter(node %in% c(53, 22, 20, 74, 9)) %>% head(3) %>% .[, c("id", "site", "node", "time")] %>% as.data.frame() -->
<!-- ``` -->
<!-- ::: -->
<!-- ::: -->
<!-- ::: -->
</section>
<section id="lets-take-the-model-from-earlier" class="slide level2">
<h2>Let’s take the <code>model</code> from earlier…</h2>

<img data-src="popgen2023_files/figure-revealjs/unnamed-chunk-32-1.png" width="960" class="r-stretch quarto-figure-center"></section>
<section id="and-simulate-data-from-it" class="slide level2">
<h2>… and simulate data from it</h2>
<p><br></p>
<p>In our script we’ll have something like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36" data-code-line-numbers="1-3|5-6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1"></a><span class="sc">&lt;</span>... <span class="fu">population</span>() definitions ...<span class="sc">&gt;</span></span>
<span id="cb36-2"><a href="#cb36-2"></a></span>
<span id="cb36-3"><a href="#cb36-3"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(...)</span>
<span id="cb36-4"><a href="#cb36-4"></a>  </span>
<span id="cb36-5"><a href="#cb36-5"></a>ts <span class="ot">&lt;-</span></span>
<span id="cb36-6"><a href="#cb36-6"></a>  <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">50e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="and-simulate-data-from-it-1" class="slide level2">
<h2>… and simulate data from it</h2>
<p><br></p>
<p>In our script we’ll have something like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37" data-code-line-numbers="7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1"></a><span class="sc">&lt;</span>... <span class="fu">population</span>() definitions ...<span class="sc">&gt;</span></span>
<span id="cb37-2"><a href="#cb37-2"></a></span>
<span id="cb37-3"><a href="#cb37-3"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(...)</span>
<span id="cb37-4"><a href="#cb37-4"></a>  </span>
<span id="cb37-5"><a href="#cb37-5"></a>ts <span class="ot">&lt;-</span></span>
<span id="cb37-6"><a href="#cb37-6"></a>  <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">50e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span></span>
<span id="cb37-7"><a href="#cb37-7"></a>  <span class="fu">ts_mutate</span>(<span class="at">mutation_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment">
<p><br></p>
<center>
<h2>
How do we work with this <code>ts</code> thing?
</h2>
</center>
</div>
<!-- ## Conversion to other genotype formats -->
<!-- If you have a tree-sequence object `ts`, you can do... -->
<!-- ```{r} -->
<!-- #| eval: false -->
<!-- ts_vcf(ts, path = "path/to/a/file.vcf.gz") -->
<!-- ``` -->
<!-- ```{r} -->
<!-- #| eval: false -->
<!-- ts_eigenstrat(ts, prefix = "path/to/eigenstrat/prefix") -->
<!-- ``` -->
<!-- ```{r} -->
<!-- #| echo: false -->
<!-- ts <- ts_simplify(ts, simplify_to = c("pop1_1", "pop2_1")) -->
<!-- ``` -->
<!-- ```{r} -->
<!-- ts_genotypes(ts) -->
<!-- ``` -->
</section>
<section id="slendrs-r-interface-to-tskit-statistics" class="slide level2">
<h2><em>slendr</em>’s R interface to <a href="https://tskit.dev/tskit"><em>tskit</em></a> statistics</h2>
<center>
<p><img data-src="images/slendr_tskit.png"></p>
</center>
<p>Allele-frequecy spectrum, nucleotide diversity <span class="math inline">\(\pi\)</span>, <span class="math inline">\(F_{ST}\)</span>, etc.</p>
</section>
<section id="extracting-sample-information" class="slide level2">
<h2>Extracting sample information</h2>
<p>We can get samples recorded in <code>ts</code> with <code>ts_samples()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 3
  name    time pop  
  &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;
1 pop1_1     0 pop1 
2 pop1_2     0 pop1 
3 pop1_3     0 pop1 
4 pop1_4     0 pop1 
5 pop1_5     0 pop1 </code></pre>
</div>
</div>
<div class="fragment">
<p>Sometimes a shortcut <code>ts_names()</code> can be useful:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_names</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pop1_1" "pop1_2" "pop1_3" "pop1_4" "pop1_5"</code></pre>
</div>
</div>
</div>
</section>
<section id="example-allele-frequency-spectrum" class="slide level2">
<h2>Example: <a href="https://en.wikipedia.org/wiki/Allele_frequency_spectrum">allele frequency spectrum</a></h2>
<div class="columns">
<div class="column" style="width:45%;">
<p>Sample 5 individuals:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>names <span class="ot">&lt;-</span> <span class="fu">ts_names</span>(ts)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>names</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pop_1" "pop_2" "pop_3" "pop_4" "pop_5"</code></pre>
</div>
</div>
<div class="fragment">
<p><br></p>
<p>Compute the AFS:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>afs <span class="ot">&lt;-</span> <span class="fu">ts_afs</span>(ts, <span class="fu">list</span>(names))</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>afs[<span class="sc">-</span><span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 4181 2078 1349 1043  838  636  616  509  461  451</code></pre>
</div>
</div>
</div>
</div><div class="column" style="width:2%;">
<p>&nbsp;</p>
</div><div class="column" style="width:53%;">
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(afs[<span class="sc">-</span><span class="dv">1</span>], <span class="at">type =</span> <span class="st">"b"</span>,</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"allele count bin"</span>,</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">"frequency"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output-display">
<p><img data-src="popgen2023_files/figure-revealjs/unnamed-chunk-45-1.png" width="960"></p>
</div>
</div>
</div>
</div>
</div>
</section></section>
<section>
<section id="exercise-2" class="title-slide slide level1 center">
<h1>Exercise #2</h1>
<p><br><br><br><br></p>

<aside><div>
<p>Use <a href="https://bodkan.quarto.pub/ku-popgen2023-onepage/">these</a> one-page handouts for reference.</p>
</div></aside></section>
<section id="exercise-2-estimating-n_e-from-afs" class="slide level2">
<h2>Exercise #2: estimating <span class="math inline">\(N_e\)</span> from AFS</h2>
<p>Imagine you sequenced 10 samples from a population and computed this AFS vector (# singletons, doubletons, etc.):</p>
<!-- ```{r} -->
<!-- #| echo: false -->
<!-- dput(as.vector(observed_afs)) -->
<!-- ``` -->
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>afs_observed <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">2520</span>, <span class="dv">1449</span>, <span class="dv">855</span>, <span class="dv">622</span>, <span class="dv">530</span>, <span class="dv">446</span>, <span class="dv">365</span>, <span class="dv">334</span>, <span class="dv">349</span>, <span class="dv">244</span>,</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>                  <span class="dv">264</span>, <span class="dv">218</span>,  <span class="dv">133</span>, <span class="dv">173</span>, <span class="dv">159</span>, <span class="dv">142</span>, <span class="dv">167</span>, <span class="dv">129</span>, <span class="dv">125</span>, <span class="dv">143</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment">
<p>You know that the population had constant <span class="math inline">\(N_e\)</span> somewhere between 1000 and 30000 for the past 100,000 generations, and had mutation and recombination rates of 1e-8.</p>
</div>
<div class="fragment">
<hr>
<p><strong>Try to guess the true </strong> <span class="math inline">\(N_e\)</span> given the observed AFS by running single-population simulations for a range of <span class="math inline">\(N_e\)</span> values and comparing <code>ts_afs()</code> results to <code>afs_observed</code>.</p>
</div>
</section>
<section id="exercise-2-hints" class="slide level2">
<h2>Exercise #2: hints</h2>
<ol type="1">
<li>Write a function (in a new script <code>model2.R</code>) taking <code>Ne</code> as a parameter. It should create one population, compile a model, run <code>ts_afs()</code> for 10 samples, and return the AFS.</li>
</ol>
<div class="fragment">
<p>If we call it <code>simulate_afs()</code>, it should work like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">simulate_afs</span>(<span class="at">Ne =</span> <span class="dv">1000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 4000 2000 1333 1000  800  667  571  500  444  400  364</code></pre>
</div>
</div>
</div>
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">simulate_afs</span>(<span class="at">Ne =</span> <span class="dv">12000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 48000 24000 16000 12000  9600  8000  6857  6000  5333  4800  4364</code></pre>
</div>
</div>
</div>

<aside><div>
<p>[Don’t panic—you can pretty much just tweak bits of my earlier example code.]</p>
</div></aside></section>
<section id="exercise-2-hints-1" class="slide level2">
<h2>Exercise #2: hints</h2>
<ol start="2" type="1">
<li><p>Using your custom <code>simulate_afs()</code> function, find the value of <code>Ne</code> that will give the closest AFS to the observed AFS:</p>
<ul>
<li><p><em>option #1</em> [<em>easy</em>]: Plot AFS vectors for various <span class="math inline">\(N_e\)</span> values, then eyeball which looks closest to the observed AFS</p></li>
<li><p><em>option #2</em> [<em>harder</em>]: Simulate AFS vectors in steps of possible <code>Ne</code> (maybe <code>lapply()</code>?), find the <a href="https://en.wikipedia.org/wiki/Mean_squared_error">closest</a> AFS</p></li>
</ul></li>
</ol>
</section>
<section id="exercise-2-hints-2" class="slide level2">
<h2>Exercise #2: hints</h2>
<p>If you’re completely lost, you can start from this “template” again:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr)</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a><span class="fu">init_env</span>()</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>simulate_afs <span class="ot">&lt;-</span> <span class="cf">function</span>(Ne) {</span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>  <span class="sc">&lt;</span>... AFS simulation code you cannibalized from my slides ...<span class="sc">&gt;</span></span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a>afs_1k  <span class="ot">&lt;-</span> <span class="fu">simulate_afs</span>(<span class="at">Ne =</span> <span class="dv">1000</span>)</span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a>afs_12k <span class="ot">&lt;-</span> <span class="fu">simulate_afs</span>(<span class="at">Ne =</span> <span class="dv">12000</span>)</span>
<span id="cb52-10"><a href="#cb52-10" aria-hidden="true" tabindex="-1"></a><span class="sc">&lt;</span>...<span class="sc">&gt;</span></span>
<span id="cb52-11"><a href="#cb52-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-12"><a href="#cb52-12" aria-hidden="true" tabindex="-1"></a><span class="er">&lt;</span>... some plotting or statistics ...<span class="sc">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section>
<section id="exercise-2-solution" class="title-slide slide level1 center">
<h1>Exercise #2: solution</h1>
<p><br><br><br><br></p>

<aside><div>
<p>See <a href="https://github.com/bodkan/ku-popgen2023/blob/main/solutions/ex2_simple.R"><code>ex2_simple.R</code></a> for a simple “eyeballing” solution.<br></p>
<p>See <a href="https://github.com/bodkan/ku-popgen2023/blob/main/solutions/ex2_grid.R"><code>ex2_grid.R</code></a> for a more elaborate grid-search solution.</p>
</div></aside></section>

<section>
<section id="gene-flow-admixture" class="title-slide slide level1 center">
<h1>Gene flow / admixture</h1>

</section>
<section id="we-use-the-function-gene_flow" class="slide level2">
<h2>We use the function <code>gene_flow()</code></h2>
<p><br></p>
<p>If we have populations <code>p1</code> and <code>p2</code>, we schedule gene flow as:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">&lt;-</span> <span class="fu">gene_flow</span>(<span class="at">from =</span> p1, <span class="at">to =</span> p2, <span class="at">start =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">600</span>, <span class="at">rate =</span> <span class="fl">0.13</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment">
<p><br></p>
<p>Multiple gene-flow events can be gathered in a list:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gene_flow</span>(<span class="at">from =</span> p1, <span class="at">to =</span> p2, <span class="at">start =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">600</span>, <span class="at">rate =</span> <span class="fl">0.13</span>),</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gene_flow</span>(<span class="at">from =</span> ..., <span class="at">to =</span> ..., <span class="at">start =</span> ..., <span class="at">end =</span> ..., <span class="at">rate =</span> ...),</span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>  ...</span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="fragment">
<p><br></p>
<center>
<strong><code>gene_flow()</code> checks for consistency!</strong>
</center>
</div>
</section>
<section id="plugging-gene-flows-into-a-model" class="slide level2">
<h2>Plugging gene flow(s) into a model</h2>
<p><br> If you have a <em>slendr</em> script, you can do this: <br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55" data-code-line-numbers="3-7|9-13|11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1"></a><span class="sc">&lt;</span>... <span class="fu">population</span>() definitions ...<span class="sc">&gt;</span></span>
<span id="cb55-2"><a href="#cb55-2"></a></span>
<span id="cb55-3"><a href="#cb55-3"></a>gf <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb55-4"><a href="#cb55-4"></a>  <span class="fu">gene_flow</span>(<span class="at">from =</span> ... <span class="at">to =</span> ..., <span class="at">start =</span> ..., <span class="at">end =</span> ..., <span class="at">rate =</span> ...),</span>
<span id="cb55-5"><a href="#cb55-5"></a>  <span class="fu">gene_flow</span>(<span class="at">from =</span> ..., <span class="at">to =</span> ..., <span class="at">start =</span> ..., <span class="at">end =</span> ..., <span class="at">rate =</span> ...),</span>
<span id="cb55-6"><a href="#cb55-6"></a>  ...</span>
<span id="cb55-7"><a href="#cb55-7"></a>)</span>
<span id="cb55-8"><a href="#cb55-8"></a></span>
<span id="cb55-9"><a href="#cb55-9"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb55-10"><a href="#cb55-10"></a>  <span class="at">populations =</span> <span class="fu">list</span>(...),</span>
<span id="cb55-11"><a href="#cb55-11"></a>  <span class="at">gene_flow =</span> gf,           <span class="co"># &lt;--- gene flow(s) specified</span></span>
<span id="cb55-12"><a href="#cb55-12"></a>  <span class="at">generation_time =</span> <span class="sc">&lt;</span>...<span class="sc">&gt;</span></span>
<span id="cb55-13"><a href="#cb55-13"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section>
<section>
<section id="exercise-3" class="title-slide slide level1 center">
<h1>Exercise #3</h1>
<p><br><br><br><br></p>

<aside><div>
<p>Use <a href="https://bodkan.quarto.pub/ku-popgen2023-onepage/">these</a> one-page handouts for reference.</p>
</div></aside></section>
<section id="exercise-3-adding-gene-flow" class="slide level2">
<h2>Exercise #3: adding gene flow</h2>
<p>Add NEA → EUR introgression to your <code>model1.R</code> script:</p>

<img data-src="popgen2023_files/figure-revealjs/unnamed-chunk-55-1.png" width="960" class="r-stretch quarto-figure-center"><p>&nbsp;</p>
</section>
<section id="exercise-3-adding-gene-flow-1" class="slide level2">
<h2>Exercise #3: adding gene flow</h2>
<p>Add NEA → EUR introgression to your <code>model1.R</code> script:</p>

<img data-src="popgen2023_files/figure-revealjs/unnamed-chunk-56-1.png" width="960" class="r-stretch quarto-figure-center"><p>3% pulse (<code>rate = 0.03</code>) between 55-50 kya (<code>start</code>-<code>end</code>)</p>
</section></section>
<section id="exercise-3-solution" class="title-slide slide level1 center">
<h1>Exercise #3: solution</h1>
<p><br><br><br><br></p>

<aside><div>
<p>See <a href="https://github.com/bodkan/ku-popgen2023/blob/main/solutions/ex3_ex4.R"><code>ex3_ex4.R</code> script on GitHub</a> for a solution.</p>
</div></aside></section>

<section>
<section id="exercise-4" class="title-slide slide level1 center">
<h1>Exercise #4</h1>
<p><br><br><br><br></p>

<aside><div>
<p>Use <a href="https://bodkan.quarto.pub/ku-popgen2023-onepage/">these</a> one-page handouts for reference.</p>
</div></aside></section>
<section id="exercise-4-more-statistics" class="slide level2">
<h2>Exercise #4: more statistics!</h2>
<p>Practice computing tree seqence statistics on the <code>ts</code> object you simulated in your <code>model1.R</code> script.</p>

<img data-src="popgen2023_files/figure-revealjs/unnamed-chunk-57-1.png" width="672" class="r-stretch quarto-figure-center"></section>
<section id="exercise-4-more-statistics-1" class="slide level2">
<h2>Exercise #4: more statistics!</h2>
<p><strong>Compute (some of) these on your tree sequence:</strong></p>
<ul>
<li><p>nucleotide <a href="https://www.slendr.net/reference/ts_diversity.html#ref-examples"><code>ts_diversity()</code></a> in each population <br></p></li>
<li><p><a href="https://www.slendr.net/reference/ts_divergence.html#ref-examples"><code>ts_divergence()</code></a> between populations<br></p></li>
<li><p><a href="https://www.slendr.net/reference/ts_fst.html#ref-examples"><code>ts_fst()</code></a> between populations<br></p></li>
<li><p>outgroup <a href="https://www.slendr.net/reference/ts_f4ratio.html#ref-examples"><code>ts_f3()</code></a> using CHIMP as the outgroup (A!) for different pairs of “B” and “C” populations<br></p></li>
<li><p><a href="https://www.slendr.net/reference/ts_f4ratio.html#ref-examples"><code>ts_f4()</code></a> test of Neanderthal introgression in EUR</p>
<p>[hint: <span class="math inline">\(f_4\)</span>(&lt;AFR&gt;, &lt;EUR&gt;; &lt;NEAND&gt;, &lt;CHIMP&gt;)]</p></li>
</ul>
<center>
<strong>Do results match expectations given your model?</strong>
</center>
</section>
<section id="exercise-4-hint" class="slide level2">
<h2>Exercise #4: hint</h2>
<p>For multipopulation statistics, you need a (named) list of samples in each population.</p>
<p><code>ts_names()</code> has a <code>split = "pop"</code> option just for this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">ts_names</span>(ts, <span class="at">split =</span> <span class="st">"pop"</span>)</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>samples</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>$AFR
[1] "AFR_1" "AFR_2" "AFR_3"

$NEA
[1] "NEA_1" "NEA_2" "NEA_3"</code></pre>
</div>
</div>
<p>You can use this in place of <code>sample_names =</code> in code like:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_diversity</span>(ts, <span class="at">sample_sets =</span> samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section>
<section>
<section id="exercise-4-solution" class="title-slide slide level1 center">
<h1>Exercise #4: solution</h1>
<p><br><br><br><br></p>

<aside><div>
<p>See <a href="https://github.com/bodkan/ku-popgen2023/blob/main/solutions/ex3_ex4.R"><code>ex3_ex4.R</code> script on GitHub</a> for a solution.</p>
</div></aside></section>
<section id="more-information" class="slide level2">
<h2>More information</h2>
<p><br></p>
<ul>
<li><p><em>slendr</em> paper is now in <a href="https://evolbiol.peercommunityin.org/articles/rec?id=630">PCI EvolBiol</a></p></li>
<li><p>documentation, tutorials is <a href="https://www.slendr.net">here</a></p></li>
<li><p>GitHub repo (bug reports!) is <a href="http://github.com/bodkan/slendr">here</a></p></li>
</ul>
<hr>
<ul>
<li>check out my new <a href="http://github.com/bodkan/demografr"><em>demografr</em></a> inference package</li>
</ul>
<hr>
<ul>
<li>contact details at <a href="https://bodkan.net">bodkan.net</a></li>
</ul>
</section></section>
<section id="section-8" class="title-slide slide level1 center">
<h1></h1>
<h1>
<p><font color="orange">B</font><font color="darkblue">O</font><font color="green">N</font><font color="red">U</font><font color="pink">S</font> <font color="purple">H</font><font color="orange">O</font><font color="green">M</font><font color="lightblue">E</font><font color="cyan">W</font><font color="green">O</font><font color="red">R</font><font color="orange">K</font></p>
</h1></section>

<section>
<section id="time-series-data" class="title-slide slide level1 center">
<h1>Time-series data</h1>

</section>
<section id="sampling-adna-samples-through-time" class="slide level2">
<h2>Sampling aDNA samples through time</h2>
<p><br></p>
<p>Imagine we have <code>pop1</code>, <code>pop2</code>, … compiled in a <code>model</code>.</p>
<p><br></p>
<p>To record <em>ancient</em> individuals in the tree sequence, we can use <code>schedule_sampling()</code> like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="fu">schedule_sampling</span>(</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>  model,                <span class="co"># compiled slendr model object</span></span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">times =</span> <span class="fu">c</span>(<span class="dv">100</span>, <span class="dv">500</span>),  <span class="co"># at these times (can be also a single number) ...</span></span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(pop1, <span class="dv">42</span>),       <span class="co"># ... sample 42 individuals from pop1</span></span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(pop2, <span class="dv">10</span>),       <span class="co"># ... sample 10 individuals from pop2</span></span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(pop3, <span class="dv">1</span>)         <span class="co"># ... sample 1 individual from pop 3</span></span>
<span id="cb59-7"><a href="#cb59-7" aria-hidden="true" tabindex="-1"></a>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="sampling-schedule-format" class="slide level2">
<h2>Sampling schedule format</h2>
<p>The output of <code>schedule_sampling()</code> is a plain data frame:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="fu">c</span>(<span class="dv">40000</span>, <span class="dv">30000</span>, <span class="dv">20000</span>, <span class="dv">10000</span>), <span class="fu">list</span>(eur, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 7
   time pop       n y_orig x_orig y     x    
  &lt;int&gt; &lt;chr&gt; &lt;int&gt; &lt;lgl&gt;  &lt;lgl&gt;  &lt;lgl&gt; &lt;lgl&gt;
1 10000 EUR       1 NA     NA     NA    NA   
2 20000 EUR       1 NA     NA     NA    NA   
3 30000 EUR       1 NA     NA     NA    NA   
4 40000 EUR       1 NA     NA     NA    NA   </code></pre>
</div>
</div>
<div class="fragment">
<p>We can bind multiple sampling schedules together, giving us finer control about sampling:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>eur_samples <span class="ot">&lt;-</span> <span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="fu">c</span>(<span class="dv">40000</span>, <span class="dv">30000</span>, <span class="dv">20000</span>, <span class="dv">10000</span>, <span class="dv">0</span>), <span class="fu">list</span>(eur, <span class="dv">1</span>))</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>afr_samples <span class="ot">&lt;-</span> <span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="dv">0</span>, <span class="fu">list</span>(afr, <span class="dv">1</span>))</span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(eur_samples, afr_samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="how-to-use-a-sampling-schedule" class="slide level2">
<h2>How to use a sampling schedule?</h2>
<p>To sample individuals based on a given schedule, we use the <code>samples =</code> argument of the <code>msprime()</code> functions:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">samples =</span> samples, <span class="at">sequence_length =</span> <span class="fl">1e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_mutate</span>(<span class="at">mutation_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment">
<p>We can verify that only specific individuals are recorded:</p>
<div class="columns">
<div class="column" style="width:40%;">
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  name   time pop  
  &lt;chr&gt; &lt;dbl&gt; &lt;chr&gt;
1 EUR_1 40000 EUR  
2 EUR_2 30000 EUR  
3 EUR_3 20000 EUR  
4 EUR_4 10000 EUR  
5 AFR_1     0 AFR  
6 EUR_5     0 EUR  </code></pre>
</div>
</div>
</div><div class="column" style="width:60%;">
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>ts</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>╔═══════════════════════════╗
║TreeSequence               ║
╠═══════════════╤═══════════╣
║Trees          │       1119║
╟───────────────┼───────────╢
║Sequence Length│    1000000║
╟───────────────┼───────────╢
║Time Units     │generations║
╟───────────────┼───────────╢
║Sample Nodes   │         12║
╟───────────────┼───────────╢
║Total Size     │  249.4 KiB║
╚═══════════════╧═══════════╝
╔═══════════╤════╤═════════╤════════════╗
║Table      │Rows│Size     │Has Metadata║
╠═══════════╪════╪═════════╪════════════╣
║Edges      │3380│105.6 KiB│          No║
╟───────────┼────┼─────────┼────────────╢
║Individuals│   6│192 Bytes│          No║
╟───────────┼────┼─────────┼────────────╢
║Migrations │   0│  8 Bytes│          No║
╟───────────┼────┼─────────┼────────────╢
║Mutations  │1497│ 54.1 KiB│          No║
╟───────────┼────┼─────────┼────────────╢
║Nodes      │ 815│ 22.3 KiB│          No║
╟───────────┼────┼─────────┼────────────╢
║Populations│   4│341 Bytes│         Yes║
╟───────────┼────┼─────────┼────────────╢
║Provenances│   2│  3.3 KiB│          No║
╟───────────┼────┼─────────┼────────────╢
║Sites      │1497│ 36.6 KiB│          No║
╚═══════════╧════╧═════════╧════════════╝</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</section></section>
<section>
<section id="exercise-5" class="title-slide slide level1 center">
<h1>Exercise #5</h1>
<p><br><br><br><br></p>

<aside><div>
<p>Use <a href="https://bodkan.quarto.pub/ku-popgen2023-onepage/">these</a> one-page handouts for reference.</p>
</div></aside></section>
<section id="exercise-5a-ancient-samples" class="slide level2">
<h2>Exercise #5a: ancient samples</h2>

<img data-src="popgen2023_files/figure-revealjs/unnamed-chunk-67-1.png" width="960" class="r-stretch"></section>
<section id="exercise-5a-ancient-samples-1" class="slide level2">
<h2>Exercise #5a: ancient samples</h2>
<p>Simulate data from your model using this sampling:</p>
<ul>
<li>one present-day CHIMP and AFR individual</li>
<li>20 present-day EUR individuals</li>
<li>1 NEA at 70 ky, 1 NEA at 40 ky</li>
<li>1 EUR every 1000 years between 50-5 kya</li>
</ul>
<p>Reminder: you can do this by:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="co"># rbind(...) together individual schedule_sampling() data frames</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span></span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">msprime</span>(model, <span class="at">samples =</span> samples, <span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span></span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_mutate</span>(<span class="at">mutation_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="exercise-5b-f_4-ratio-statistic" class="slide level2">
<h2>Exercise #5b: <span class="math inline">\(f_4\)</span>-ratio statistic</h2>
<p>Use <span class="math inline">\(f_4\)</span>-ratio statistic to replicate the <a href="https://www.pnas.org/doi/10.1073/pnas.1814338116#fig01">following figure</a>:</p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/neand_decline.png" style="width:60.0%"></p>
</figure>
</div>

<aside><div>
<p>Hint: You can compute Neanderthal ancestry for a vector of individuals <code>X</code> as <code>ts_f4ratio(ts, X = X, "NEA_1", "NEA_2", "AFR_1", "CHIMP_1")</code>.</p>
</div></aside></section></section>
<section>
<section id="exercise-5-solution" class="title-slide slide level1 center">
<h1>Exercise #5: solution</h1>
<p><br><br><br><br></p>

<aside><div>
<p>See <a href="https://github.com/bodkan/ku-popgen2023/blob/main/solutions/ex5.R"><code>ex5.R</code> script on GitHub</a> for a solution.</p>
</div></aside></section>
<section id="conversion-to-other-genotype-formats" class="slide level2">
<h2>Conversion to other genotype formats</h2>
<p>If you have a tree-sequence object <code>ts</code>, you can do…</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_vcf</span>(ts, <span class="at">path =</span> <span class="st">"path/to/a/file.vcf.gz"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_eigenstrat</span>(ts, <span class="at">prefix =</span> <span class="st">"path/to/eigenstrat/prefix"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_genotypes</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>    pos EUR_1_chr1 EUR_1_chr2 EUR_2_chr1 EUR_2_chr2 AFR_1_chr1 AFR_1_chr2
1   376          1          1          0          0          1          0
2   498          0          0          1          1          0          1
3   810          1          1          0          0          0          0
4  2274          0          0          0          0          0          1
5  2828          0          0          1          1          0          1
6  2974          0          0          0          0          0          1
7  3475          1          1          1          1          0          0
8  4013          1          1          1          1          0          0
9  4161          0          0          0          0          0          1
10 6656          0          0          0          0          0          1</code></pre>
</div>
</div>
<div class="footer footer-default">

</div>
</section></section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="popgen2023_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="popgen2023_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="popgen2023_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="popgen2023_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="popgen2023_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="popgen2023_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="popgen2023_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="popgen2023_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="popgen2023_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="popgen2023_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>